{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Módulos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "En Python, cada script o archivo de código fuente, se denominan módulos. Estos módulos, a la vez, pueden formar parte de paquetes. Un paquete, es una carpeta que contiene archivos `.py`. Por ejemplo, si guardáramos el contenido de la función para obtener codones a partir de un string, y le ponemos la extensión `py` y lo guardamos como `get_codons.py` sería un script de python, si el script está en la misma carpeta del notebook yo podría importarlo así:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import get_codons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Eso me permite reutilizar mi código. Afortunadamente python contienen módulos `built-in`, métodos integrados. Además, podemos instalar nuevos paquetes que contienen módulos con `pip`, el instalador oficial de Python, o con `conda`, el gestor de paquetes de Anaconda Inc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import calendar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   February 2022\n",
      "Mo Tu We Th Fr Sa Su\n",
      "    1  2  3  4  5  6\n",
      " 7  8  9 10 11 12 13\n",
      "14 15 16 17 18 19 20\n",
      "21 22 23 24 25 26 27\n",
      "28\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(calendar.month(2022, 2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Es posible también, abreviar los namespaces mediante un alias. Para ello, durante la importación, se asigna la palabra clave as seguida del alias con el cuál nos referiremos en el futuro a ese namespace importado:\n",
    "\n",
    "- `import modulo`\n",
    "- `import modulo as m`\n",
    "- `import paquete.modulo1 as pm`\n",
    "- `import paquete.subpaquete.modulo1 as psm`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"./imgs/pandas.png\" align=\"center\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ¿Qué es Pandas?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pandas, de *\"panel data\"*, es una biblioteca de Python que nos permite manejar tablas, también conocidas como *DataFrames*, las cuales están constituidas por tres elementos:\n",
    "\n",
    "- Los datos\n",
    "- El índice\n",
    "- Las columnas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|Estructura de pandas | Dimensiones | Analogía Excel\n",
    "| -- | -- | -- |\n",
    "|Series |1D| Columna|\n",
    "|DataFrame |2D | Hoja de cálculo|\n",
    "|Panel | 3D| Múltiples hojas de cálculo|\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Para utilizar el módulo Pandas es necesario importarlo."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "import pandas as pd\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Descargaremos mas adelante el archivo de manera remota utilizando comandos bash en Jupyter Notebooks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "! mkdir -p data\n",
    "! curl https://atgenomics-data.s3.amazonaws.com/IGC.annotation.tsv.gz -o data/IGC.annotation.tsv.gz\n",
    "\n",
    "df = pd.read_csv(\"data/IGC.annotation.tsv.gz\", sep='\\t')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O bien, podemos descargar archivos de manera remota utilizando únicamente Pandas, recuerda quitar el `#` de la siguiente celda:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.read_csv('https://atgenomics-data.s3.amazonaws.com/IGC.annotation.tsv.gz', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Uso de comandos bash en celdas de Jupyper Notebooks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Podemos utilizar comandos bash en Jupyter por celda con el comando mágico: `%%bash`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for i in {1..10}\n",
    "do\n",
    "    echo $i\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comando mágico de IPython `%ls` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 pgc288  41K Feb  8 19:27 day01.ipynb\n",
      "-rw-r--r-- 1 pgc288  793 Feb  8 22:34 Untitled1.ipynb\n",
      "-rw-r--r-- 1 pgc288  65K Feb  8 22:41 day02.ipynb\n",
      "-rw-r--r-- 1 pgc288 7.2K Feb  8 22:54 class_and_methods.ipynb\n",
      "-rw-r--r-- 1 pgc288 390K Feb 10 08:04 day03.ipynb\n"
     ]
    }
   ],
   "source": [
    "%ls -ltrh *ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Linux command with `!`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 pgc288 staff  41K Feb  8 19:27 day01.ipynb\n",
      "-rw-r--r-- 1 pgc288 staff  793 Feb  8 22:34 Untitled1.ipynb\n",
      "-rw-r--r-- 1 pgc288 staff  65K Feb  8 22:41 day02.ipynb\n",
      "-rw-r--r-- 1 pgc288 staff 7.2K Feb  8 22:54 class_and_methods.ipynb\n",
      "-rw-r--r-- 1 pgc288 staff 390K Feb 10 08:04 day03.ipynb\n",
      "This is Bash\n"
     ]
    }
   ],
   "source": [
    "! ls -ltrh *ipynb\n",
    "\n",
    "! echo \"This is Bash\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "myvar = !ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Untitled1.ipynb',\n",
       " '__pycache__',\n",
       " 'atgtools.py',\n",
       " 'class_and_methods.ipynb',\n",
       " 'data',\n",
       " 'day01.ipynb',\n",
       " 'day02.ipynb',\n",
       " 'day03.ipynb',\n",
       " 'get_codons.py',\n",
       " 'get_gc_content.py']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myvar"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Magic Commands availables per line/cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/json": {
       "cell": {
        "!": "OSMagics",
        "HTML": "Other",
        "SVG": "Other",
        "bash": "Other",
        "capture": "ExecutionMagics",
        "debug": "ExecutionMagics",
        "file": "Other",
        "html": "DisplayMagics",
        "javascript": "DisplayMagics",
        "js": "DisplayMagics",
        "latex": "DisplayMagics",
        "markdown": "DisplayMagics",
        "perl": "Other",
        "prun": "ExecutionMagics",
        "pypy": "Other",
        "python": "Other",
        "python2": "Other",
        "python3": "Other",
        "ruby": "Other",
        "script": "ScriptMagics",
        "sh": "Other",
        "svg": "DisplayMagics",
        "sx": "OSMagics",
        "system": "OSMagics",
        "time": "ExecutionMagics",
        "timeit": "ExecutionMagics",
        "writefile": "OSMagics"
       },
       "line": {
        "alias": "OSMagics",
        "alias_magic": "BasicMagics",
        "autoawait": "AsyncMagics",
        "autocall": "AutoMagics",
        "automagic": "AutoMagics",
        "autosave": "KernelMagics",
        "bookmark": "OSMagics",
        "cat": "Other",
        "cd": "OSMagics",
        "clear": "KernelMagics",
        "colors": "BasicMagics",
        "conda": "PackagingMagics",
        "config": "ConfigMagics",
        "connect_info": "KernelMagics",
        "cp": "Other",
        "debug": "ExecutionMagics",
        "dhist": "OSMagics",
        "dirs": "OSMagics",
        "doctest_mode": "BasicMagics",
        "ed": "Other",
        "edit": "KernelMagics",
        "env": "OSMagics",
        "gui": "BasicMagics",
        "hist": "Other",
        "history": "HistoryMagics",
        "killbgscripts": "ScriptMagics",
        "ldir": "Other",
        "less": "KernelMagics",
        "lf": "Other",
        "lk": "Other",
        "ll": "Other",
        "load": "CodeMagics",
        "load_ext": "ExtensionMagics",
        "loadpy": "CodeMagics",
        "logoff": "LoggingMagics",
        "logon": "LoggingMagics",
        "logstart": "LoggingMagics",
        "logstate": "LoggingMagics",
        "logstop": "LoggingMagics",
        "ls": "Other",
        "lsmagic": "BasicMagics",
        "lx": "Other",
        "macro": "ExecutionMagics",
        "magic": "BasicMagics",
        "man": "KernelMagics",
        "matplotlib": "PylabMagics",
        "mkdir": "Other",
        "more": "KernelMagics",
        "mv": "Other",
        "notebook": "BasicMagics",
        "page": "BasicMagics",
        "pastebin": "CodeMagics",
        "pdb": "ExecutionMagics",
        "pdef": "NamespaceMagics",
        "pdoc": "NamespaceMagics",
        "pfile": "NamespaceMagics",
        "pinfo": "NamespaceMagics",
        "pinfo2": "NamespaceMagics",
        "pip": "PackagingMagics",
        "popd": "OSMagics",
        "pprint": "BasicMagics",
        "precision": "BasicMagics",
        "prun": "ExecutionMagics",
        "psearch": "NamespaceMagics",
        "psource": "NamespaceMagics",
        "pushd": "OSMagics",
        "pwd": "OSMagics",
        "pycat": "OSMagics",
        "pylab": "PylabMagics",
        "qtconsole": "KernelMagics",
        "quickref": "BasicMagics",
        "recall": "HistoryMagics",
        "rehashx": "OSMagics",
        "reload_ext": "ExtensionMagics",
        "rep": "Other",
        "rerun": "HistoryMagics",
        "reset": "NamespaceMagics",
        "reset_selective": "NamespaceMagics",
        "rm": "Other",
        "rmdir": "Other",
        "run": "ExecutionMagics",
        "save": "CodeMagics",
        "sc": "OSMagics",
        "set_env": "OSMagics",
        "store": "StoreMagics",
        "sx": "OSMagics",
        "system": "OSMagics",
        "tb": "ExecutionMagics",
        "time": "ExecutionMagics",
        "timeit": "ExecutionMagics",
        "unalias": "OSMagics",
        "unload_ext": "ExtensionMagics",
        "who": "NamespaceMagics",
        "who_ls": "NamespaceMagics",
        "whos": "NamespaceMagics",
        "xdel": "NamespaceMagics",
        "xmode": "BasicMagics"
       }
      },
      "text/plain": [
       "Available line magics:\n",
       "%alias  %alias_magic  %autoawait  %autocall  %automagic  %autosave  %bookmark  %cat  %cd  %clear  %colors  %conda  %config  %connect_info  %cp  %debug  %dhist  %dirs  %doctest_mode  %ed  %edit  %env  %gui  %hist  %history  %killbgscripts  %ldir  %less  %lf  %lk  %ll  %load  %load_ext  %loadpy  %logoff  %logon  %logstart  %logstate  %logstop  %ls  %lsmagic  %lx  %macro  %magic  %man  %matplotlib  %mkdir  %more  %mv  %notebook  %page  %pastebin  %pdb  %pdef  %pdoc  %pfile  %pinfo  %pinfo2  %pip  %popd  %pprint  %precision  %prun  %psearch  %psource  %pushd  %pwd  %pycat  %pylab  %qtconsole  %quickref  %recall  %rehashx  %reload_ext  %rep  %rerun  %reset  %reset_selective  %rm  %rmdir  %run  %save  %sc  %set_env  %store  %sx  %system  %tb  %time  %timeit  %unalias  %unload_ext  %who  %who_ls  %whos  %xdel  %xmode\n",
       "\n",
       "Available cell magics:\n",
       "%%!  %%HTML  %%SVG  %%bash  %%capture  %%debug  %%file  %%html  %%javascript  %%js  %%latex  %%markdown  %%perl  %%prun  %%pypy  %%python  %%python2  %%python3  %%ruby  %%script  %%sh  %%svg  %%sx  %%system  %%time  %%timeit  %%writefile\n",
       "\n",
       "Automagic is ON, % prefix IS NOT needed for line magics."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%lsmagic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Uso de únicamente Python para descargar archivos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Descargar el archivo de manera remota utilizando ÚNICAMENTE módulos de Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import requests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Crear directorio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs('./data/', exist_ok = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Definir variable con URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://atgenomics-data.s3.amazonaws.com/IGC.annotation.tsv.gz'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Utilizar requests para obtener el contenido del archivo en la URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = requests.get(url, allow_redirects=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Escribir el contenido de request a un archivo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30375588"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "open('./data/IGC.annotation.tsv.gz', 'wb').write(r.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cargamos el paquete Pandas y  archivo a pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/IGC.annotation.tsv.gz', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>24615</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Weissella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.066298</td>\n",
       "      <td>0.072897</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.095501</td>\n",
       "      <td>0.096262</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "0  911104.WcibK1_010100007220        24615          Complete            SP   \n",
       "1            585054.EFER_0542        21669          Complete            SP   \n",
       "2          SZEY-48A_GL0052647        20778          Complete           CHN   \n",
       "3     1048689.ECO55CA74_02930        20778          Complete            SP   \n",
       "4            MH0427_GL0087973        20775          Complete           EUR   \n",
       "\n",
       "  Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "0                         Firmicutes                         Weissella   \n",
       "1                     Proteobacteria                       Escherichia   \n",
       "2                     Proteobacteria                       Escherichia   \n",
       "3                     Proteobacteria                       Escherichia   \n",
       "4                     Proteobacteria                       Escherichia   \n",
       "\n",
       "  KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "0         unknown          NOG12793                    0.066298   \n",
       "1         unknown          NOG12793                    0.058406   \n",
       "2          K01317          NOG12793                    0.203631   \n",
       "3          K01317          NOG12793                    0.095501   \n",
       "4          K01317          NOG12793                    0.098658   \n",
       "\n",
       "   Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "0                        0.072897                    unknown   \n",
       "1                        0.062617                    unknown   \n",
       "2                        0.212150            Enzyme Families   \n",
       "3                        0.096262            Enzyme Families   \n",
       "4                        0.104673            Enzyme Families   \n",
       "\n",
       "  eggNOG Functional Categories Cohort Assembled  \n",
       "0             Function unknown              EUR  \n",
       "1             Function unknown          EUR;CHN  \n",
       "2             Function unknown          EUR;CHN  \n",
       "3             Function unknown          EUR;CHN  \n",
       "4             Function unknown      EUR;CHN;USA  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Leer archivos en formato Excel, *xlsx*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Podemos leer un archivo Excel directamente e imprimir el nombre de las hojas de cálculo del archivo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# xlsx = pd.ExcelFile('miarchivoExcel.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lista el nombre de todas las hojas de cálculo:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# xlsx.sheet_names\n",
    "\n",
    "# df = pd.read_excel(xlsx, \"Datos\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O bien, leer directamente la hoja de cálculo definiento el nombre, el número de filas que tiene que omitir -generalmente la cabecera en muchos documentos oficiales- y la columna que definiremos como el índice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.read_excel('miarchivoExcel.xlsx', sheet_name='Datos', skiprows=3, index_col=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Descripción general de *DataFrames*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Es importante recordar siempre qué tipo de dato estamos utilizando en cada línea. En este caso, `df` es un:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Obtenemos las dimensiones del DataFrame, el número de filas y columnas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1285642"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1285642, 13)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "¿Qué tipo de datos contiene cada celda? ¿Es así? La asignación de tipo de dato la hace pandas de manera automática, es importante corregir algunos tipos de datos si queremos utilizar ciertos métodos. Por ejemplo, no podemos usar métodos numéricos donde hay valores de 'objetos' (strings,listas, diccionarios); tampoco podemos aplicar métodos numéricos en una columna que tiene sólo valores categóricos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene Name                              object\n",
       "Gene Length                             int64\n",
       "Gene Completeness                      object\n",
       "Cohort Origin                          object\n",
       "Taxonomic Annotation(Phylum Level)     object\n",
       "Taxonomic Annotation(Genus Level)      object\n",
       "KEGG Annotation                        object\n",
       "eggNOG Annotation                      object\n",
       "Sample Occurence Frequency            float64\n",
       "Individual Occurence Frequency        float64\n",
       "KEGG Functional Categories             object\n",
       "eggNOG Functional Categories           object\n",
       "Cohort Assembled                       object\n",
       "dtype: object"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|Pandas dtype| Python | Uso|\n",
    "|--|--|--|\n",
    "|`object`| string o mixto| Texto o números en string, lista o diccionario|\n",
    "|`int64`|int | Números enteros|\n",
    "|`float64`|float | Números decimales|\n",
    "|`bool`|bool| Valores True/False|\n",
    "|`datetime64`|datetime | Valores de fechas y hora|\n",
    "|`category`| NA | Categorías en texto"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Una versión extendida descriptiva del DataFrame. AL final vemos el uso de memoria de este grupo de datos, ¿es poco, mucho? ¿Qué creen que sea el `Non-Null Count`?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1285642 entries, 0 to 1285641\n",
      "Data columns (total 13 columns):\n",
      " #   Column                              Non-Null Count    Dtype  \n",
      "---  ------                              --------------    -----  \n",
      " 0   Gene Name                           1285642 non-null  object \n",
      " 1   Gene Length                         1285642 non-null  int64  \n",
      " 2   Gene Completeness                   1285642 non-null  object \n",
      " 3   Cohort Origin                       1285642 non-null  object \n",
      " 4   Taxonomic Annotation(Phylum Level)  1285642 non-null  object \n",
      " 5   Taxonomic Annotation(Genus Level)   1285642 non-null  object \n",
      " 6   KEGG Annotation                     1285642 non-null  object \n",
      " 7   eggNOG Annotation                   1285642 non-null  object \n",
      " 8   Sample Occurence Frequency          1285642 non-null  float64\n",
      " 9   Individual Occurence Frequency      1285642 non-null  float64\n",
      " 10  KEGG Functional Categories          1285642 non-null  object \n",
      " 11  eggNOG Functional Categories        1285642 non-null  object \n",
      " 12  Cohort Assembled                    1124785 non-null  object \n",
      "dtypes: float64(2), int64(1), object(10)\n",
      "memory usage: 127.5+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "El método `df.head()` nos permite obtener las primeras 5 líneas del DataFrame. Podemos definir el número deseado en el en medio de los paréntesis, por ejemplo `df.head(20)` para obtener las primeras 20 líneas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>24615</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Weissella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.066298</td>\n",
       "      <td>0.072897</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.095501</td>\n",
       "      <td>0.096262</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "0  911104.WcibK1_010100007220        24615          Complete            SP   \n",
       "1            585054.EFER_0542        21669          Complete            SP   \n",
       "2          SZEY-48A_GL0052647        20778          Complete           CHN   \n",
       "3     1048689.ECO55CA74_02930        20778          Complete            SP   \n",
       "4            MH0427_GL0087973        20775          Complete           EUR   \n",
       "\n",
       "  Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "0                         Firmicutes                         Weissella   \n",
       "1                     Proteobacteria                       Escherichia   \n",
       "2                     Proteobacteria                       Escherichia   \n",
       "3                     Proteobacteria                       Escherichia   \n",
       "4                     Proteobacteria                       Escherichia   \n",
       "\n",
       "  KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "0         unknown          NOG12793                    0.066298   \n",
       "1         unknown          NOG12793                    0.058406   \n",
       "2          K01317          NOG12793                    0.203631   \n",
       "3          K01317          NOG12793                    0.095501   \n",
       "4          K01317          NOG12793                    0.098658   \n",
       "\n",
       "   Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "0                        0.072897                    unknown   \n",
       "1                        0.062617                    unknown   \n",
       "2                        0.212150            Enzyme Families   \n",
       "3                        0.096262            Enzyme Families   \n",
       "4                        0.104673            Enzyme Families   \n",
       "\n",
       "  eggNOG Functional Categories Cohort Assembled  \n",
       "0             Function unknown              EUR  \n",
       "1             Function unknown          EUR;CHN  \n",
       "2             Function unknown          EUR;CHN  \n",
       "3             Function unknown          EUR;CHN  \n",
       "4             Function unknown      EUR;CHN;USA  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ahora las últimas 5."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1285637</th>\n",
       "      <td>665954.HMPREF1017_02554</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.089187</td>\n",
       "      <td>0.077570</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285638</th>\n",
       "      <td>753642.ECNC101_09444</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285639</th>\n",
       "      <td>768724.HMPREF9131_0965</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Peptoniphilus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285640</th>\n",
       "      <td>866771.HMPREF9296_1721</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285641</th>\n",
       "      <td>926026.ECO5101_18612</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>proNOG05789</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "1285637  665954.HMPREF1017_02554          102          Complete            SP   \n",
       "1285638     753642.ECNC101_09444          102          Complete            SP   \n",
       "1285639   768724.HMPREF9131_0965          102          Complete            SP   \n",
       "1285640   866771.HMPREF9296_1721          102          Complete            SP   \n",
       "1285641     926026.ECO5101_18612          102          Complete            SP   \n",
       "\n",
       "        Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "1285637                      Bacteroidetes                       Bacteroides   \n",
       "1285638                     Proteobacteria                       Escherichia   \n",
       "1285639                         Firmicutes                     Peptoniphilus   \n",
       "1285640                      Bacteroidetes                        Prevotella   \n",
       "1285641                     Proteobacteria                       Escherichia   \n",
       "\n",
       "        KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "1285637         unknown           unknown                    0.089187   \n",
       "1285638         unknown           unknown                    0.001579   \n",
       "1285639         unknown           unknown                    0.000789   \n",
       "1285640         unknown           unknown                    0.000789   \n",
       "1285641         unknown       proNOG05789                    0.001579   \n",
       "\n",
       "         Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "1285637                        0.077570                    unknown   \n",
       "1285638                        0.000935                    unknown   \n",
       "1285639                        0.000935                    unknown   \n",
       "1285640                        0.000935                    unknown   \n",
       "1285641                        0.001869                    unknown   \n",
       "\n",
       "        eggNOG Functional Categories Cohort Assembled  \n",
       "1285637                      unknown              NaN  \n",
       "1285638                      unknown              NaN  \n",
       "1285639                      unknown              NaN  \n",
       "1285640                      unknown              NaN  \n",
       "1285641             Function unknown              NaN  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Podemos usar `sample()` para muestrear aleatoriamente filas del DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3724</th>\n",
       "      <td>V1.UC31-0_GL0030911</td>\n",
       "      <td>4548</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>K00284</td>\n",
       "      <td>COG0067;COG0069;COG0070</td>\n",
       "      <td>0.342541</td>\n",
       "      <td>0.348598</td>\n",
       "      <td>Energy Metabolism</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "3724  V1.UC31-0_GL0030911         4548          Complete           EUR   \n",
       "\n",
       "     Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "3724                         Firmicutes                           unknown   \n",
       "\n",
       "     KEGG Annotation        eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "3724          K00284  COG0067;COG0069;COG0070                    0.342541   \n",
       "\n",
       "      Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "3724                        0.348598          Energy Metabolism   \n",
       "\n",
       "             eggNOG Functional Categories Cohort Assembled  \n",
       "3724  Amino acid transport and metabolism      EUR;CHN;USA  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sample()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Podemor colocar el puntero en medio de los paréntesis de `df.sample()` y utilizar las teclas `SHIFT + TAB` para obtener ayuda del método de pandas, o de cualquier método que estés aplicando:\n",
    "\n",
    "![green-divider](imgs/help.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1125162</th>\n",
       "      <td>MH0434_GL0244560</td>\n",
       "      <td>285</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>unknown</td>\n",
       "      <td>COG3436</td>\n",
       "      <td>0.128650</td>\n",
       "      <td>0.137383</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1168023</th>\n",
       "      <td>MH0011_GL0004548</td>\n",
       "      <td>240</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG14713</td>\n",
       "      <td>0.209155</td>\n",
       "      <td>0.201869</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>737000</th>\n",
       "      <td>MH0161_GL0014740</td>\n",
       "      <td>711</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>K01104</td>\n",
       "      <td>COG4464</td>\n",
       "      <td>0.399369</td>\n",
       "      <td>0.384112</td>\n",
       "      <td>Cellular Processes and Signaling</td>\n",
       "      <td>Carbohydrate transport and metabolism;Cell wal...</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>944007</th>\n",
       "      <td>MH0266_GL0022060</td>\n",
       "      <td>474</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>K00783</td>\n",
       "      <td>COG1576</td>\n",
       "      <td>0.002368</td>\n",
       "      <td>0.002804</td>\n",
       "      <td>Poorly Characterized</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945462</th>\n",
       "      <td>T2D-63A_GL0026446</td>\n",
       "      <td>474</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>K06950</td>\n",
       "      <td>COG1418</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>Poorly Characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1213317</th>\n",
       "      <td>159733294-stool1_revised_scaffold26513_1_gene1...</td>\n",
       "      <td>198</td>\n",
       "      <td>Complete</td>\n",
       "      <td>USA</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112079</th>\n",
       "      <td>V1.UC35-0_GL0110065</td>\n",
       "      <td>1866</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Porphyromonas</td>\n",
       "      <td>K00174</td>\n",
       "      <td>COG0674;COG1014</td>\n",
       "      <td>0.009471</td>\n",
       "      <td>0.008411</td>\n",
       "      <td>Carbohydrate Metabolism</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313806</th>\n",
       "      <td>MH0356_GL0008889</td>\n",
       "      <td>1269</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Bifidobacterium</td>\n",
       "      <td>K01436</td>\n",
       "      <td>COG1473</td>\n",
       "      <td>0.270718</td>\n",
       "      <td>0.267290</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235063</th>\n",
       "      <td>MH0057_GL0019817</td>\n",
       "      <td>1407</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Paraprevotella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG75428</td>\n",
       "      <td>0.192581</td>\n",
       "      <td>0.178505</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139796</th>\n",
       "      <td>MH0274_GL0144378</td>\n",
       "      <td>1722</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG235407</td>\n",
       "      <td>0.080505</td>\n",
       "      <td>0.084112</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Gene Name  Gene Length  \\\n",
       "1125162                                   MH0434_GL0244560          285   \n",
       "1168023                                   MH0011_GL0004548          240   \n",
       "737000                                    MH0161_GL0014740          711   \n",
       "944007                                    MH0266_GL0022060          474   \n",
       "945462                                   T2D-63A_GL0026446          474   \n",
       "1213317  159733294-stool1_revised_scaffold26513_1_gene1...          198   \n",
       "112079                                 V1.UC35-0_GL0110065         1866   \n",
       "313806                                    MH0356_GL0008889         1269   \n",
       "235063                                    MH0057_GL0019817         1407   \n",
       "139796                                    MH0274_GL0144378         1722   \n",
       "\n",
       "        Gene Completeness Cohort Origin Taxonomic Annotation(Phylum Level)  \\\n",
       "1125162          Complete           EUR                      Bacteroidetes   \n",
       "1168023          Complete           EUR                         Firmicutes   \n",
       "737000           Complete           EUR                         Firmicutes   \n",
       "944007           Complete           EUR                      Bacteroidetes   \n",
       "945462           Complete           CHN                         Firmicutes   \n",
       "1213317          Complete           USA                      Bacteroidetes   \n",
       "112079           Complete           EUR                      Bacteroidetes   \n",
       "313806           Complete           EUR                     Actinobacteria   \n",
       "235063           Complete           EUR                      Bacteroidetes   \n",
       "139796           Complete           EUR                      Bacteroidetes   \n",
       "\n",
       "        Taxonomic Annotation(Genus Level) KEGG Annotation eggNOG Annotation  \\\n",
       "1125162                       Bacteroides         unknown           COG3436   \n",
       "1168023                  Faecalibacterium         unknown          NOG14713   \n",
       "737000                   Faecalibacterium          K01104           COG4464   \n",
       "944007                            unknown          K00783           COG1576   \n",
       "945462                            unknown          K06950           COG1418   \n",
       "1213317                           unknown         unknown           unknown   \n",
       "112079                      Porphyromonas          K00174   COG0674;COG1014   \n",
       "313806                    Bifidobacterium          K01436           COG1473   \n",
       "235063                     Paraprevotella         unknown          NOG75428   \n",
       "139796                            unknown         unknown         NOG235407   \n",
       "\n",
       "         Sample Occurence Frequency  Individual Occurence Frequency  \\\n",
       "1125162                    0.128650                        0.137383   \n",
       "1168023                    0.209155                        0.201869   \n",
       "737000                     0.399369                        0.384112   \n",
       "944007                     0.002368                        0.002804   \n",
       "945462                     0.001579                        0.001869   \n",
       "1213317                    0.000789                        0.000935   \n",
       "112079                     0.009471                        0.008411   \n",
       "313806                     0.270718                        0.267290   \n",
       "235063                     0.192581                        0.178505   \n",
       "139796                     0.080505                        0.084112   \n",
       "\n",
       "               KEGG Functional Categories  \\\n",
       "1125162                           unknown   \n",
       "1168023                           unknown   \n",
       "737000   Cellular Processes and Signaling   \n",
       "944007               Poorly Characterized   \n",
       "945462               Poorly Characterized   \n",
       "1213317                           unknown   \n",
       "112079            Carbohydrate Metabolism   \n",
       "313806                    Enzyme Families   \n",
       "235063                            unknown   \n",
       "139796                            unknown   \n",
       "\n",
       "                              eggNOG Functional Categories Cohort Assembled  \n",
       "1125162              Replication, recombination and repair              EUR  \n",
       "1168023                                   Function unknown          EUR;USA  \n",
       "737000   Carbohydrate transport and metabolism;Cell wal...      EUR;CHN;USA  \n",
       "944007                                    Function unknown          EUR;CHN  \n",
       "945462                    General function prediction only              CHN  \n",
       "1213317                                            unknown              USA  \n",
       "112079                    Energy production and conversion              EUR  \n",
       "313806                    General function prediction only          EUR;CHN  \n",
       "235063                                             unknown      EUR;CHN;USA  \n",
       "139796                                    Function unknown      EUR;CHN;USA  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cuando manejamos columnas numéricas podemos utilizar `describe()` para obtener algunas medidas de tendencia central. Es importante notar que si una columna de valores booleanos o categóricos en código numérico tienen una etiqueta numérica con `df.dtypes` se obtendrá los descriptores aunque estos carezcan de sentido categórico."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.285642e+06</td>\n",
       "      <td>1.285642e+06</td>\n",
       "      <td>1.285642e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>9.581168e+02</td>\n",
       "      <td>1.361281e-01</td>\n",
       "      <td>1.352274e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>7.285365e+02</td>\n",
       "      <td>1.941363e-01</td>\n",
       "      <td>1.932935e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.020000e+02</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>4.530000e+02</td>\n",
       "      <td>1.420679e-02</td>\n",
       "      <td>1.401869e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>8.190000e+02</td>\n",
       "      <td>4.814522e-02</td>\n",
       "      <td>4.859813e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.257000e+03</td>\n",
       "      <td>1.752170e-01</td>\n",
       "      <td>1.738318e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.461500e+04</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene Length  Sample Occurence Frequency  \\\n",
       "count  1.285642e+06                1.285642e+06   \n",
       "mean   9.581168e+02                1.361281e-01   \n",
       "std    7.285365e+02                1.941363e-01   \n",
       "min    1.020000e+02                0.000000e+00   \n",
       "25%    4.530000e+02                1.420679e-02   \n",
       "50%    8.190000e+02                4.814522e-02   \n",
       "75%    1.257000e+03                1.752170e-01   \n",
       "max    2.461500e+04                1.000000e+00   \n",
       "\n",
       "       Individual Occurence Frequency  \n",
       "count                    1.285642e+06  \n",
       "mean                     1.352274e-01  \n",
       "std                      1.932935e-01  \n",
       "min                      0.000000e+00  \n",
       "25%                      1.401869e-02  \n",
       "50%                      4.859813e-02  \n",
       "75%                      1.738318e-01  \n",
       "max                      1.000000e+00  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Podemos trasponer una tabla con el método `transpose()` o la propiedad `T`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene Length</th>\n",
       "      <td>1285642.0</td>\n",
       "      <td>958.116782</td>\n",
       "      <td>728.536549</td>\n",
       "      <td>102.0</td>\n",
       "      <td>453.000000</td>\n",
       "      <td>819.000000</td>\n",
       "      <td>1257.000000</td>\n",
       "      <td>24615.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <td>1285642.0</td>\n",
       "      <td>0.136128</td>\n",
       "      <td>0.194136</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014207</td>\n",
       "      <td>0.048145</td>\n",
       "      <td>0.175217</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <td>1285642.0</td>\n",
       "      <td>0.135227</td>\n",
       "      <td>0.193294</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014019</td>\n",
       "      <td>0.048598</td>\n",
       "      <td>0.173832</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    count        mean         std    min  \\\n",
       "Gene Length                     1285642.0  958.116782  728.536549  102.0   \n",
       "Sample Occurence Frequency      1285642.0    0.136128    0.194136    0.0   \n",
       "Individual Occurence Frequency  1285642.0    0.135227    0.193294    0.0   \n",
       "\n",
       "                                       25%         50%          75%      max  \n",
       "Gene Length                     453.000000  819.000000  1257.000000  24615.0  \n",
       "Sample Occurence Frequency        0.014207    0.048145     0.175217      1.0  \n",
       "Individual Occurence Frequency    0.014019    0.048598     0.173832      1.0  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe().transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene Length</th>\n",
       "      <td>1285642.0</td>\n",
       "      <td>958.116782</td>\n",
       "      <td>728.536549</td>\n",
       "      <td>102.0</td>\n",
       "      <td>453.000000</td>\n",
       "      <td>819.000000</td>\n",
       "      <td>1257.000000</td>\n",
       "      <td>24615.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <td>1285642.0</td>\n",
       "      <td>0.136128</td>\n",
       "      <td>0.194136</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014207</td>\n",
       "      <td>0.048145</td>\n",
       "      <td>0.175217</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <td>1285642.0</td>\n",
       "      <td>0.135227</td>\n",
       "      <td>0.193294</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014019</td>\n",
       "      <td>0.048598</td>\n",
       "      <td>0.173832</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    count        mean         std    min  \\\n",
       "Gene Length                     1285642.0  958.116782  728.536549  102.0   \n",
       "Sample Occurence Frequency      1285642.0    0.136128    0.194136    0.0   \n",
       "Individual Occurence Frequency  1285642.0    0.135227    0.193294    0.0   \n",
       "\n",
       "                                       25%         50%          75%      max  \n",
       "Gene Length                     453.000000  819.000000  1257.000000  24615.0  \n",
       "Sample Occurence Frequency        0.014207    0.048145     0.175217      1.0  \n",
       "Individual Occurence Frequency    0.014019    0.048598     0.173832      1.0  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe().T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "En Python, al ser un lenguaje de tipado dinámico, podemos definir una variable en cualquier momento sin definirla al principio. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "description = df.describe().T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "y guardar el contenido de una variable, o de un DataFrame, a un archivo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "',count,mean,std,min,25%,50%,75%,max\\nGene Length,1285642.0,958.116782121306,728.536548632577,102.0,453.0,819.0,1257.0,24615.0\\nSample Occurence Frequency,1285642.0,0.13612813347528085,0.19413631329975223,0.0,0.0142067876874507,0.0481452249408051,0.175217048145225,1.0\\nIndividual Occurence Frequency,1285642.0,0.13522741836228966,0.1932935302121455,0.0,0.014018691588785,0.0485981308411215,0.173831775700935,1.0\\n'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "description.to_csv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "description.to_csv('data/description.tsv', sep='\\t', na_rep=\"**\", index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To Copy or not To Copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_test = pd.DataFrame({'A': [1, 2, 3]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   A\n",
       "0  1\n",
       "1  2\n",
       "2  3"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sub = df_test[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   A\n",
       "0  1\n",
       "1  2"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/6w/ckq52h85633bflldtk3cffpr0000gn/T/ipykernel_57591/3394583580.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_sub.loc[0, 'A'] = -1\n"
     ]
    }
   ],
   "source": [
    "df_sub.loc[0, 'A'] = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   A\n",
       "0 -1\n",
       "1  2"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   A\n",
       "0 -1\n",
       "1  2\n",
       "2  3"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Una modificación de un subset modifica el DataFrame original, para evitar este posible problema podemos usar el método `copy()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_test = pd.DataFrame({'A': [1, 2, 3]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_copy = df_test[0:2].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_copy.loc[0, 'A'] = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   A\n",
       "0 -1\n",
       "1  2"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   A\n",
       "0  1\n",
       "1  2\n",
       "2  3"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Podemos usar un **encadenamiento de métodos**, una técnica que se utiliza para realizar varias llamadas a métodos en el mismo objeto, utilizando la referencia del objeto solo una vez: `df.uno().dos().tres().cuatro()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene Length</th>\n",
       "      <td>100.0</td>\n",
       "      <td>14855.250000</td>\n",
       "      <td>2779.328970</td>\n",
       "      <td>12138.000000</td>\n",
       "      <td>12709.500000</td>\n",
       "      <td>13570.500000</td>\n",
       "      <td>16718.250000</td>\n",
       "      <td>24615.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <td>100.0</td>\n",
       "      <td>0.162928</td>\n",
       "      <td>0.162151</td>\n",
       "      <td>0.002368</td>\n",
       "      <td>0.059984</td>\n",
       "      <td>0.093528</td>\n",
       "      <td>0.206393</td>\n",
       "      <td>0.831097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <td>100.0</td>\n",
       "      <td>0.163607</td>\n",
       "      <td>0.159957</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>0.061682</td>\n",
       "      <td>0.095794</td>\n",
       "      <td>0.213084</td>\n",
       "      <td>0.831776</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                count          mean          std  \\\n",
       "Gene Length                     100.0  14855.250000  2779.328970   \n",
       "Sample Occurence Frequency      100.0      0.162928     0.162151   \n",
       "Individual Occurence Frequency  100.0      0.163607     0.159957   \n",
       "\n",
       "                                         min           25%           50%  \\\n",
       "Gene Length                     12138.000000  12709.500000  13570.500000   \n",
       "Sample Occurence Frequency          0.002368      0.059984      0.093528   \n",
       "Individual Occurence Frequency      0.001869      0.061682      0.095794   \n",
       "\n",
       "                                         75%           max  \n",
       "Gene Length                     16718.250000  24615.000000  \n",
       "Sample Occurence Frequency          0.206393      0.831097  \n",
       "Individual Occurence Frequency      0.213084      0.831776  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(100).describe().T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "El atributo `columns` y `values` permite accedeer a las columnas y a los valores del DataFrame, respectivamente."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Name\n",
      "Gene Length\n",
      "Gene Completeness\n",
      "Cohort Origin\n",
      "Taxonomic Annotation(Phylum Level)\n",
      "Taxonomic Annotation(Genus Level)\n",
      "KEGG Annotation\n",
      "eggNOG Annotation\n",
      "Sample Occurence Frequency\n",
      "Individual Occurence Frequency\n",
      "KEGG Functional Categories\n",
      "eggNOG Functional Categories\n",
      "Cohort Assembled\n"
     ]
    }
   ],
   "source": [
    "for x in df.columns:\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'list' object has no attribute 'upper'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Input \u001b[0;32mIn [51]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43m[\u001b[49m\u001b[43mx\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mx\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mupper\u001b[49m()\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'list' object has no attribute 'upper'"
     ]
    }
   ],
   "source": [
    "[x for x in df.columns].upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Index' object has no attribute 'upper'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Input \u001b[0;32mIn [52]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mupper\u001b[49m()\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Index' object has no attribute 'upper'"
     ]
    }
   ],
   "source": [
    "df.columns.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['GENE NAME', 'GENE LENGTH', 'GENE COMPLETENESS', 'COHORT ORIGIN',\n",
       "       'TAXONOMIC ANNOTATION(PHYLUM LEVEL)',\n",
       "       'TAXONOMIC ANNOTATION(GENUS LEVEL)', 'KEGG ANNOTATION',\n",
       "       'EGGNOG ANNOTATION', 'SAMPLE OCCURENCE FREQUENCY',\n",
       "       'INDIVIDUAL OCCURENCE FREQUENCY', 'KEGG FUNCTIONAL CATEGORIES',\n",
       "       'EGGNOG FUNCTIONAL CATEGORIES', 'COHORT ASSEMBLED'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns.str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['911104.WcibK1_010100007220', 24615, 'Complete', ..., 'unknown',\n",
       "        'Function unknown', 'EUR'],\n",
       "       ['585054.EFER_0542', 21669, 'Complete', ..., 'unknown',\n",
       "        'Function unknown', 'EUR;CHN'],\n",
       "       ['SZEY-48A_GL0052647', 20778, 'Complete', ..., 'Enzyme Families',\n",
       "        'Function unknown', 'EUR;CHN'],\n",
       "       ...,\n",
       "       ['768724.HMPREF9131_0965', 102, 'Complete', ..., 'unknown',\n",
       "        'unknown', nan],\n",
       "       ['866771.HMPREF9296_1721', 102, 'Complete', ..., 'unknown',\n",
       "        'unknown', nan],\n",
       "       ['926026.ECO5101_18612', 102, 'Complete', ..., 'unknown',\n",
       "        'Function unknown', nan]], dtype=object)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "¿Qué tipo de datos son los siguientes atributos?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tuple"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.series.Series"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df.dtypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.indexes.base.Index"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "numpy.ndarray"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Selección de columnas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Seleccionamos una columna con la sintaxis `[]` para una sola columna o `[[]]` para una lista de columnas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene Name', 'Gene Length', 'Gene Completeness', 'Cohort Origin',\n",
       "       'Taxonomic Annotation(Phylum Level)',\n",
       "       'Taxonomic Annotation(Genus Level)', 'KEGG Annotation',\n",
       "       'eggNOG Annotation', 'Sample Occurence Frequency',\n",
       "       'Individual Occurence Frequency', 'KEGG Functional Categories',\n",
       "       'eggNOG Functional Categories', 'Cohort Assembled'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'Gene'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Input \u001b[0;32mIn [60]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mGene\u001b[49m\n",
      "File \u001b[0;32m~/bin/miniconda3/lib/python3.9/site-packages/pandas/core/generic.py:5583\u001b[0m, in \u001b[0;36mNDFrame.__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5576\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m   5577\u001b[0m     name \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_internal_names_set\n\u001b[1;32m   5578\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m name \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_metadata\n\u001b[1;32m   5579\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m name \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_accessors\n\u001b[1;32m   5580\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_info_axis\u001b[38;5;241m.\u001b[39m_can_hold_identifiers_and_holds_name(name)\n\u001b[1;32m   5581\u001b[0m ):\n\u001b[1;32m   5582\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m[name]\n\u001b[0;32m-> 5583\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mobject\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__getattribute__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'Gene'"
     ]
    }
   ],
   "source": [
    "df.Gene "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df['Gene Name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    911104.WcibK1_010100007220\n",
       "1              585054.EFER_0542\n",
       "2            SZEY-48A_GL0052647\n",
       "3       1048689.ECO55CA74_02930\n",
       "4              MH0427_GL0087973\n",
       "Name: Gene Name, dtype: object"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Name'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>24615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene Name  Gene Length\n",
       "0  911104.WcibK1_010100007220        24615\n",
       "1            585054.EFER_0542        21669\n",
       "2          SZEY-48A_GL0052647        20778\n",
       "3     1048689.ECO55CA74_02930        20778\n",
       "4            MH0427_GL0087973        20775"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Gene Name', 'Gene Length']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.series.Series"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df['Gene Name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df[['Gene Name']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df[['Gene Name', 'Gene Length']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cuando transponemos las columnas el índice ahora ocupa el valor de las columnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene Name</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Length</th>\n",
       "      <td>24615</td>\n",
       "      <td>21669</td>\n",
       "      <td>20778</td>\n",
       "      <td>20778</td>\n",
       "      <td>20775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      0                 1                   2  \\\n",
       "Gene Name    911104.WcibK1_010100007220  585054.EFER_0542  SZEY-48A_GL0052647   \n",
       "Gene Length                       24615             21669               20778   \n",
       "\n",
       "                                   3                 4  \n",
       "Gene Name    1048689.ECO55CA74_02930  MH0427_GL0087973  \n",
       "Gene Length                    20778             20775  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Gene Name', 'Gene Length']].head().T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Con `set_index()` definimos la columna que usaremos como índice del DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = df.head().set_index('Gene Length')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>24615</td>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Weissella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.066298</td>\n",
       "      <td>0.072897</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21669</td>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20778</td>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20778</td>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.095501</td>\n",
       "      <td>0.096262</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20775</td>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene Length                   Gene Name Gene Completeness Cohort Origin  \\\n",
       "0        24615  911104.WcibK1_010100007220          Complete            SP   \n",
       "1        21669            585054.EFER_0542          Complete            SP   \n",
       "2        20778          SZEY-48A_GL0052647          Complete           CHN   \n",
       "3        20778     1048689.ECO55CA74_02930          Complete            SP   \n",
       "4        20775            MH0427_GL0087973          Complete           EUR   \n",
       "\n",
       "  Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "0                         Firmicutes                         Weissella   \n",
       "1                     Proteobacteria                       Escherichia   \n",
       "2                     Proteobacteria                       Escherichia   \n",
       "3                     Proteobacteria                       Escherichia   \n",
       "4                     Proteobacteria                       Escherichia   \n",
       "\n",
       "  KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "0         unknown          NOG12793                    0.066298   \n",
       "1         unknown          NOG12793                    0.058406   \n",
       "2          K01317          NOG12793                    0.203631   \n",
       "3          K01317          NOG12793                    0.095501   \n",
       "4          K01317          NOG12793                    0.098658   \n",
       "\n",
       "   Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "0                        0.072897                    unknown   \n",
       "1                        0.062617                    unknown   \n",
       "2                        0.212150            Enzyme Families   \n",
       "3                        0.096262            Enzyme Families   \n",
       "4                        0.104673            Enzyme Families   \n",
       "\n",
       "  eggNOG Functional Categories Cohort Assembled  \n",
       "0             Function unknown              EUR  \n",
       "1             Function unknown          EUR;CHN  \n",
       "2             Function unknown          EUR;CHN  \n",
       "3             Function unknown          EUR;CHN  \n",
       "4             Function unknown      EUR;CHN;USA  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>24615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene Name  Gene Length\n",
       "0  911104.WcibK1_010100007220        24615\n",
       "1            585054.EFER_0542        21669\n",
       "2          SZEY-48A_GL0052647        20778\n",
       "3     1048689.ECO55CA74_02930        20778\n",
       "4            MH0427_GL0087973        20775"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Gene Name', 'Gene Length']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene Name</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Length</th>\n",
       "      <td>24615</td>\n",
       "      <td>21669</td>\n",
       "      <td>20778</td>\n",
       "      <td>20778</td>\n",
       "      <td>20775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      0                 1                   2  \\\n",
       "Gene Name    911104.WcibK1_010100007220  585054.EFER_0542  SZEY-48A_GL0052647   \n",
       "Gene Length                       24615             21669               20778   \n",
       "\n",
       "                                   3                 4  \n",
       "Gene Name    1048689.ECO55CA74_02930  MH0427_GL0087973  \n",
       "Gene Length                    20778             20775  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Gene Name', 'Gene Length']].head().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>911104.WcibK1_010100007220</th>\n",
       "      <td>24615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585054.EFER_0542</th>\n",
       "      <td>21669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZEY-48A_GL0052647</th>\n",
       "      <td>20778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048689.ECO55CA74_02930</th>\n",
       "      <td>20778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MH0427_GL0087973</th>\n",
       "      <td>20775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Gene Length\n",
       "Gene Name                              \n",
       "911104.WcibK1_010100007220        24615\n",
       "585054.EFER_0542                  21669\n",
       "SZEY-48A_GL0052647                20778\n",
       "1048689.ECO55CA74_02930           20778\n",
       "MH0427_GL0087973                  20775"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Gene Name', 'Gene Length']].set_index('Gene Name').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mientras no definamos el resultado a una nueva variable o sobreescribamos el dataframe original el nuevo índice no se guardará."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Name</th>\n",
       "      <th>911104.WcibK1_010100007220</th>\n",
       "      <th>585054.EFER_0542</th>\n",
       "      <th>SZEY-48A_GL0052647</th>\n",
       "      <th>1048689.ECO55CA74_02930</th>\n",
       "      <th>MH0427_GL0087973</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene Length</th>\n",
       "      <td>24615</td>\n",
       "      <td>21669</td>\n",
       "      <td>20778</td>\n",
       "      <td>20778</td>\n",
       "      <td>20775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Name    911104.WcibK1_010100007220  585054.EFER_0542  SZEY-48A_GL0052647  \\\n",
       "Gene Length                       24615             21669               20778   \n",
       "\n",
       "Gene Name    1048689.ECO55CA74_02930  MH0427_GL0087973  \n",
       "Gene Length                    20778             20775  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Gene Name', 'Gene Length']].set_index('Gene Name').head().T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ordenamiento de filas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>24615</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Weissella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.066298</td>\n",
       "      <td>0.072897</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.095501</td>\n",
       "      <td>0.096262</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "0  911104.WcibK1_010100007220        24615          Complete            SP   \n",
       "1            585054.EFER_0542        21669          Complete            SP   \n",
       "2          SZEY-48A_GL0052647        20778          Complete           CHN   \n",
       "3     1048689.ECO55CA74_02930        20778          Complete            SP   \n",
       "4            MH0427_GL0087973        20775          Complete           EUR   \n",
       "\n",
       "  Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "0                         Firmicutes                         Weissella   \n",
       "1                     Proteobacteria                       Escherichia   \n",
       "2                     Proteobacteria                       Escherichia   \n",
       "3                     Proteobacteria                       Escherichia   \n",
       "4                     Proteobacteria                       Escherichia   \n",
       "\n",
       "  KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "0         unknown          NOG12793                    0.066298   \n",
       "1         unknown          NOG12793                    0.058406   \n",
       "2          K01317          NOG12793                    0.203631   \n",
       "3          K01317          NOG12793                    0.095501   \n",
       "4          K01317          NOG12793                    0.098658   \n",
       "\n",
       "   Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "0                        0.072897                    unknown   \n",
       "1                        0.062617                    unknown   \n",
       "2                        0.212150            Enzyme Families   \n",
       "3                        0.096262            Enzyme Families   \n",
       "4                        0.104673            Enzyme Families   \n",
       "\n",
       "  eggNOG Functional Categories Cohort Assembled  \n",
       "0             Function unknown              EUR  \n",
       "1             Function unknown          EUR;CHN  \n",
       "2             Function unknown          EUR;CHN  \n",
       "3             Function unknown          EUR;CHN  \n",
       "4             Function unknown      EUR;CHN;USA  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/6w/ckq52h85633bflldtk3cffpr0000gn/T/ipykernel_57591/3578354586.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.sort_index will be keyword-only.\n",
      "  df.sort_index('Gene Length').head()\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "No axis named Gene Length for object type DataFrame",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/bin/miniconda3/lib/python3.9/site-packages/pandas/core/generic.py:550\u001b[0m, in \u001b[0;36mNDFrame._get_axis_number\u001b[0;34m(cls, axis)\u001b[0m\n\u001b[1;32m    549\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 550\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mcls\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_AXIS_TO_AXIS_NUMBER\u001b[49m\u001b[43m[\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m    551\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m:\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Gene Length'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [74]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msort_index\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mGene Length\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mhead()\n",
      "File \u001b[0;32m~/bin/miniconda3/lib/python3.9/site-packages/pandas/util/_decorators.py:311\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    305\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    306\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    307\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39marguments),\n\u001b[1;32m    308\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    309\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mstacklevel,\n\u001b[1;32m    310\u001b[0m     )\n\u001b[0;32m--> 311\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/bin/miniconda3/lib/python3.9/site-packages/pandas/core/frame.py:6449\u001b[0m, in \u001b[0;36mDataFrame.sort_index\u001b[0;34m(self, axis, level, ascending, inplace, kind, na_position, sort_remaining, ignore_index, key)\u001b[0m\n\u001b[1;32m   6347\u001b[0m \u001b[38;5;129m@deprecate_nonkeyword_arguments\u001b[39m(version\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, allowed_args\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mself\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m   6348\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msort_index\u001b[39m(\n\u001b[1;32m   6349\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   6358\u001b[0m     key: IndexKeyFunc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   6359\u001b[0m ):\n\u001b[1;32m   6360\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   6361\u001b[0m \u001b[38;5;124;03m    Sort object by labels (along an axis).\u001b[39;00m\n\u001b[1;32m   6362\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   6447\u001b[0m \u001b[38;5;124;03m    d  4\u001b[39;00m\n\u001b[1;32m   6448\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 6449\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msort_index\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   6450\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   6451\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   6452\u001b[0m \u001b[43m        \u001b[49m\u001b[43mascending\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   6453\u001b[0m \u001b[43m        \u001b[49m\u001b[43minplace\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   6454\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkind\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   6455\u001b[0m \u001b[43m        \u001b[49m\u001b[43mna_position\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   6456\u001b[0m \u001b[43m        \u001b[49m\u001b[43msort_remaining\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   6457\u001b[0m \u001b[43m        \u001b[49m\u001b[43mignore_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   6458\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   6459\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/bin/miniconda3/lib/python3.9/site-packages/pandas/core/generic.py:4688\u001b[0m, in \u001b[0;36mNDFrame.sort_index\u001b[0;34m(self, axis, level, ascending, inplace, kind, na_position, sort_remaining, ignore_index, key)\u001b[0m\n\u001b[1;32m   4674\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msort_index\u001b[39m(\n\u001b[1;32m   4675\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   4676\u001b[0m     axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4684\u001b[0m     key: IndexKeyFunc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   4685\u001b[0m ):\n\u001b[1;32m   4687\u001b[0m     inplace \u001b[38;5;241m=\u001b[39m validate_bool_kwarg(inplace, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minplace\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 4688\u001b[0m     axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_axis_number\u001b[49m\u001b[43m(\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4689\u001b[0m     ascending \u001b[38;5;241m=\u001b[39m validate_ascending(ascending)\n\u001b[1;32m   4691\u001b[0m     target \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_axis(axis)\n",
      "File \u001b[0;32m~/bin/miniconda3/lib/python3.9/site-packages/pandas/core/generic.py:552\u001b[0m, in \u001b[0;36mNDFrame._get_axis_number\u001b[0;34m(cls, axis)\u001b[0m\n\u001b[1;32m    550\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m_AXIS_TO_AXIS_NUMBER[axis]\n\u001b[1;32m    551\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m:\n\u001b[0;32m--> 552\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNo axis named \u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m for object type \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: No axis named Gene Length for object type DataFrame"
     ]
    }
   ],
   "source": [
    "df.sort_index('Gene Length').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cohort Assembled</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EUR</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "      <td>24615</td>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>0.072897</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.066298</td>\n",
       "      <td>Weissella</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EUR;CHN</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "      <td>21669</td>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR;CHN</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Complete</td>\n",
       "      <td>20778</td>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>K01317</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EUR;CHN</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "      <td>20778</td>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>0.096262</td>\n",
       "      <td>K01317</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>0.095501</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Complete</td>\n",
       "      <td>20775</td>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>K01317</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Cohort Assembled Cohort Origin Gene Completeness  Gene Length  \\\n",
       "0              EUR            SP          Complete        24615   \n",
       "1          EUR;CHN            SP          Complete        21669   \n",
       "2          EUR;CHN           CHN          Complete        20778   \n",
       "3          EUR;CHN            SP          Complete        20778   \n",
       "4      EUR;CHN;USA           EUR          Complete        20775   \n",
       "\n",
       "                    Gene Name  Individual Occurence Frequency KEGG Annotation  \\\n",
       "0  911104.WcibK1_010100007220                        0.072897         unknown   \n",
       "1            585054.EFER_0542                        0.062617         unknown   \n",
       "2          SZEY-48A_GL0052647                        0.212150          K01317   \n",
       "3     1048689.ECO55CA74_02930                        0.096262          K01317   \n",
       "4            MH0427_GL0087973                        0.104673          K01317   \n",
       "\n",
       "  KEGG Functional Categories  Sample Occurence Frequency  \\\n",
       "0                    unknown                    0.066298   \n",
       "1                    unknown                    0.058406   \n",
       "2            Enzyme Families                    0.203631   \n",
       "3            Enzyme Families                    0.095501   \n",
       "4            Enzyme Families                    0.098658   \n",
       "\n",
       "  Taxonomic Annotation(Genus Level) Taxonomic Annotation(Phylum Level)  \\\n",
       "0                         Weissella                         Firmicutes   \n",
       "1                       Escherichia                     Proteobacteria   \n",
       "2                       Escherichia                     Proteobacteria   \n",
       "3                       Escherichia                     Proteobacteria   \n",
       "4                       Escherichia                     Proteobacteria   \n",
       "\n",
       "  eggNOG Annotation eggNOG Functional Categories  \n",
       "0          NOG12793             Function unknown  \n",
       "1          NOG12793             Function unknown  \n",
       "2          NOG12793             Function unknown  \n",
       "3          NOG12793             Function unknown  \n",
       "4          NOG12793             Function unknown  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_index(axis=1, level='Gene Length').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>24615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene Length\n",
       "0        24615\n",
       "1        21669\n",
       "2        20778\n",
       "3        20778\n",
       "4        20775"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Gene Length']].sort_index(axis=1).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ordenamiento por columna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1285641</th>\n",
       "      <td>926026.ECO5101_18612</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>proNOG05789</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285206</th>\n",
       "      <td>O2.UC13-0_GL0011042</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.002368</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285205</th>\n",
       "      <td>O2.UC12-1_GL0011992</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Collinsella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.011839</td>\n",
       "      <td>0.010280</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285204</th>\n",
       "      <td>O2.CD3-0-PT_GL0126180</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285203</th>\n",
       "      <td>O2.CD3-0-PT_GL0020295</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridium</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.011839</td>\n",
       "      <td>0.013084</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "1285641   926026.ECO5101_18612          102          Complete            SP   \n",
       "1285206    O2.UC13-0_GL0011042          102          Complete           EUR   \n",
       "1285205    O2.UC12-1_GL0011992          102          Complete           EUR   \n",
       "1285204  O2.CD3-0-PT_GL0126180          102          Complete           EUR   \n",
       "1285203  O2.CD3-0-PT_GL0020295          102          Complete           EUR   \n",
       "\n",
       "        Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "1285641                     Proteobacteria                       Escherichia   \n",
       "1285206                     Actinobacteria                           unknown   \n",
       "1285205                     Actinobacteria                       Collinsella   \n",
       "1285204                         Firmicutes                           unknown   \n",
       "1285203                         Firmicutes                       Clostridium   \n",
       "\n",
       "        KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "1285641         unknown       proNOG05789                    0.001579   \n",
       "1285206         unknown           unknown                    0.002368   \n",
       "1285205         unknown           unknown                    0.011839   \n",
       "1285204         unknown           unknown                    0.001579   \n",
       "1285203         unknown           unknown                    0.011839   \n",
       "\n",
       "         Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "1285641                        0.001869                    unknown   \n",
       "1285206                        0.000935                    unknown   \n",
       "1285205                        0.010280                    unknown   \n",
       "1285204                        0.000935                    unknown   \n",
       "1285203                        0.013084                    unknown   \n",
       "\n",
       "        eggNOG Functional Categories Cohort Assembled  \n",
       "1285641             Function unknown              NaN  \n",
       "1285206                      unknown              EUR  \n",
       "1285205                      unknown              EUR  \n",
       "1285204                      unknown              EUR  \n",
       "1285203                      unknown              EUR  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_values('Gene Length').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>24615</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Weissella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.066298</td>\n",
       "      <td>0.072897</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.095501</td>\n",
       "      <td>0.096262</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "0  911104.WcibK1_010100007220        24615          Complete            SP   \n",
       "1            585054.EFER_0542        21669          Complete            SP   \n",
       "2          SZEY-48A_GL0052647        20778          Complete           CHN   \n",
       "3     1048689.ECO55CA74_02930        20778          Complete            SP   \n",
       "4            MH0427_GL0087973        20775          Complete           EUR   \n",
       "\n",
       "  Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "0                         Firmicutes                         Weissella   \n",
       "1                     Proteobacteria                       Escherichia   \n",
       "2                     Proteobacteria                       Escherichia   \n",
       "3                     Proteobacteria                       Escherichia   \n",
       "4                     Proteobacteria                       Escherichia   \n",
       "\n",
       "  KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "0         unknown          NOG12793                    0.066298   \n",
       "1         unknown          NOG12793                    0.058406   \n",
       "2          K01317          NOG12793                    0.203631   \n",
       "3          K01317          NOG12793                    0.095501   \n",
       "4          K01317          NOG12793                    0.098658   \n",
       "\n",
       "   Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "0                        0.072897                    unknown   \n",
       "1                        0.062617                    unknown   \n",
       "2                        0.212150            Enzyme Families   \n",
       "3                        0.096262            Enzyme Families   \n",
       "4                        0.104673            Enzyme Families   \n",
       "\n",
       "  eggNOG Functional Categories Cohort Assembled  \n",
       "0             Function unknown              EUR  \n",
       "1             Function unknown          EUR;CHN  \n",
       "2             Function unknown          EUR;CHN  \n",
       "3             Function unknown          EUR;CHN  \n",
       "4             Function unknown      EUR;CHN;USA  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_values('Gene Length', ascending=False).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Podemos seleccionar una columna, ordenar los valores de manera descendente y obtiener las primeras líneas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df.sort_values('Gene Length', ascending=False).head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.series.Series"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df['Gene Length'].sort_values(ascending=False).head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**¿Cuántas categorías funcionales KEGG hay, y cuántos genes de cada categoría?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "unknown                                                                        512851\n",
       "Carbohydrate Metabolism                                                         90254\n",
       "Membrane Transport                                                              72775\n",
       "Cellular Processes and Signaling                                                61821\n",
       "Poorly Characterized                                                            61400\n",
       "                                                                                ...  \n",
       "Cell Motility;Membrane Transport                                                    1\n",
       "Carbohydrate Metabolism;Transport and Catabolism                                    1\n",
       "Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism;Metabolism               1\n",
       "Folding, Sorting and Degradation;Genetic Information Processing;Translation         1\n",
       "Energy Metabolism;Genetic Information Processing                                    1\n",
       "Name: KEGG Functional Categories, Length: 199, dtype: int64"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['KEGG Functional Categories'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['unknown', 'Enzyme Families', 'Membrane Transport',\n",
       "       'Infectious Diseases', 'Amino Acid Metabolism',\n",
       "       'Carbohydrate Metabolism', 'Cellular Processes and Signaling',\n",
       "       'Metabolism', 'Signaling Molecules and Interaction',\n",
       "       'Cellular Processes and Signaling;Signaling Molecules and Interaction',\n",
       "       'Replication and Repair', 'Lipid Metabolism',\n",
       "       'Genetic Information Processing',\n",
       "       'Metabolism of Terpenoids and Polyketides',\n",
       "       'Glycan Biosynthesis and Metabolism',\n",
       "       'Amino Acid Metabolism;Genetic Information Processing',\n",
       "       'Membrane Transport;Signaling Molecules and Interaction',\n",
       "       'Metabolism of Other Amino Acids', 'Poorly Characterized',\n",
       "       'Signal Transduction', 'Transcription',\n",
       "       'Genetic Information Processing;Replication and Repair',\n",
       "       'Metabolism of Cofactors and Vitamins',\n",
       "       'Carbohydrate Metabolism;Metabolism', 'Translation',\n",
       "       'Nucleotide Metabolism', 'Energy Metabolism',\n",
       "       'Folding, Sorting and Degradation',\n",
       "       'Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism',\n",
       "       'Xenobiotics Biodegradation and Metabolism',\n",
       "       'Carbohydrate Metabolism;Enzyme Families', 'Cell Growth and Death',\n",
       "       'Metabolism;Signaling Molecules and Interaction',\n",
       "       'Enzyme Families;Metabolism',\n",
       "       'Cellular Processes and Signaling;Metabolism',\n",
       "       'Cellular Processes and Signaling;Poorly Characterized',\n",
       "       'Enzyme Families;Signal Transduction',\n",
       "       'Nucleotide Metabolism;Replication and Repair',\n",
       "       'Biosynthesis of Other Secondary Metabolites',\n",
       "       'Metabolism;Signal Transduction', 'Cell Motility',\n",
       "       'Cellular Processes and Signaling;Enzyme Families',\n",
       "       'Metabolism;Replication and Repair',\n",
       "       'Carbohydrate Metabolism;Poorly Characterized',\n",
       "       'Signal Transduction;Transcription',\n",
       "       'Poorly Characterized;Translation',\n",
       "       'Genetic Information Processing;Transcription',\n",
       "       'Cellular Processes and Signaling;Lipid Metabolism',\n",
       "       'Folding, Sorting and Degradation;Infectious Diseases',\n",
       "       'Infectious Diseases;Poorly Characterized',\n",
       "       'Carbohydrate Metabolism;Energy Metabolism',\n",
       "       'Amino Acid Metabolism;Infectious Diseases',\n",
       "       'Glycan Biosynthesis and Metabolism;Metabolism',\n",
       "       'Amino Acid Metabolism;Cellular Processes and Signaling',\n",
       "       'Cellular Processes and Signaling;Transcription',\n",
       "       'Transport and Catabolism',\n",
       "       'Cell Growth and Death;Signal Transduction',\n",
       "       'Cellular Processes and Signaling;Metabolism of Other Amino Acids',\n",
       "       'Cellular Processes and Signaling;Enzyme Families;Metabolism',\n",
       "       'Excretory System', 'Metabolism;Nucleotide Metabolism',\n",
       "       'Metabolism;Metabolism of Other Amino Acids',\n",
       "       'Genetic Information Processing;Metabolism',\n",
       "       'Infectious Diseases;Metabolism',\n",
       "       'Folding, Sorting and Degradation;Translation',\n",
       "       'Enzyme Families;Metabolism;Signal Transduction',\n",
       "       'Endocrine System', 'Amino Acid Metabolism;Poorly Characterized',\n",
       "       'Lipid Metabolism;Poorly Characterized',\n",
       "       'Cellular Processes and Signaling;Membrane Transport',\n",
       "       'Carbohydrate Metabolism;Lipid Metabolism',\n",
       "       'Enzyme Families;Folding, Sorting and Degradation',\n",
       "       'Glycan Biosynthesis and Metabolism;Metabolism of Terpenoids and Polyketides',\n",
       "       'Carbohydrate Metabolism;Metabolism of Terpenoids and Polyketides',\n",
       "       'Cellular Processes and Signaling;Signal Transduction',\n",
       "       'Cell Growth and Death;Metabolism',\n",
       "       'Membrane Transport;Replication and Repair',\n",
       "       'Membrane Transport;Poorly Characterized',\n",
       "       'Energy Metabolism;Folding, Sorting and Degradation;Metabolism',\n",
       "       'Energy Metabolism;Metabolism',\n",
       "       'Cellular Processes and Signaling;Infectious Diseases',\n",
       "       'Genetic Information Processing;Membrane Transport',\n",
       "       'Enzyme Families;Metabolism;unknown',\n",
       "       'Cellular Processes and Signaling;Glycan Biosynthesis and Metabolism',\n",
       "       'Amino Acid Metabolism;Metabolism',\n",
       "       'Metabolism of Cofactors and Vitamins;Poorly Characterized',\n",
       "       'Metabolism;Metabolism of Cofactors and Vitamins',\n",
       "       'Amino Acid Metabolism;Enzyme Families',\n",
       "       'Glycan Biosynthesis and Metabolism;Lipid Metabolism',\n",
       "       'Replication and Repair;Transcription', 'Immune System Diseases',\n",
       "       'Lipid Metabolism;Metabolism of Terpenoids and Polyketides',\n",
       "       'Carbohydrate Metabolism;Cellular Processes and Signaling',\n",
       "       'Enzyme Families;Transport and Catabolism',\n",
       "       'Enzyme Families;unknown', 'Cell Growth and Death;Enzyme Families',\n",
       "       'Energy Metabolism;Xenobiotics Biodegradation and Metabolism',\n",
       "       'Cellular Processes and Signaling;Nucleotide Metabolism',\n",
       "       'Amino Acid Metabolism;Energy Metabolism',\n",
       "       'Carbohydrate Metabolism;Membrane Transport',\n",
       "       'Biosynthesis of Other Secondary Metabolites;Lipid Metabolism',\n",
       "       'Cellular Processes and Signaling;Replication and Repair',\n",
       "       'Metabolism of Other Amino Acids;Xenobiotics Biodegradation and Metabolism',\n",
       "       'Carbohydrate Metabolism;Metabolism of Cofactors and Vitamins',\n",
       "       'Cellular Processes and Signaling;unknown',\n",
       "       'Nucleotide Metabolism;Poorly Characterized', 'Metabolism;unknown',\n",
       "       'Carbohydrate Metabolism;Nucleotide Metabolism',\n",
       "       'Metabolism of Other Amino Acids;Transcription',\n",
       "       'Replication and Repair;Signal Transduction',\n",
       "       'Energy Metabolism;Nucleotide Metabolism',\n",
       "       'Cellular Processes and Signaling;Folding, Sorting and Degradation',\n",
       "       'Carbohydrate Metabolism;Transcription',\n",
       "       'Amino Acid Metabolism;Carbohydrate Metabolism',\n",
       "       'Enzyme Families;Replication and Repair',\n",
       "       'Carbohydrate Metabolism;Signaling Molecules and Interaction',\n",
       "       'Lipid Metabolism;Replication and Repair',\n",
       "       'Enzyme Families;Signaling Molecules and Interaction',\n",
       "       'Folding, Sorting and Degradation;Membrane Transport',\n",
       "       'Lipid Metabolism;Metabolism',\n",
       "       'Metabolism;Metabolism of Terpenoids and Polyketides',\n",
       "       'Folding, Sorting and Degradation;Genetic Information Processing',\n",
       "       'Carbohydrate Metabolism;Replication and Repair',\n",
       "       'Metabolism of Cofactors and Vitamins;Xenobiotics Biodegradation and Metabolism',\n",
       "       'Folding, Sorting and Degradation;Genetic Information Processing;Translation',\n",
       "       'Metabolism;Poorly Characterized',\n",
       "       'Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism;Metabolism',\n",
       "       'Carbohydrate Metabolism;Transport and Catabolism',\n",
       "       'Cell Motility;Membrane Transport',\n",
       "       'Folding, Sorting and Degradation;Metabolism',\n",
       "       'Lipid Metabolism;Xenobiotics Biodegradation and Metabolism',\n",
       "       'Metabolism;Translation', 'Metabolism;Transcription',\n",
       "       'Lipid Metabolism;Transcription',\n",
       "       'Poorly Characterized;Signal Transduction',\n",
       "       'Carbohydrate Metabolism;Cellular Processes and Signaling;Metabolism',\n",
       "       'Metabolism;Xenobiotics Biodegradation and Metabolism',\n",
       "       'Carbohydrate Metabolism;Signal Transduction',\n",
       "       'Cell Motility;Poorly Characterized',\n",
       "       'Amino Acid Metabolism;Xenobiotics Biodegradation and Metabolism',\n",
       "       'Folding, Sorting and Degradation;Metabolism of Cofactors and Vitamins',\n",
       "       'Amino Acid Metabolism;Transcription',\n",
       "       'Membrane Transport;Signal Transduction',\n",
       "       'Membrane Transport;Transcription',\n",
       "       'Amino Acid Metabolism;Metabolism of Cofactors and Vitamins',\n",
       "       'Metabolism of Terpenoids and Polyketides;Translation',\n",
       "       'Poorly Characterized;Xenobiotics Biodegradation and Metabolism',\n",
       "       'Metabolism of Cofactors and Vitamins;Metabolism of Other Amino Acids',\n",
       "       'Biosynthesis of Other Secondary Metabolites;Xenobiotics Biodegradation and Metabolism',\n",
       "       'Carbohydrate Metabolism;Metabolism;Metabolism of Terpenoids and Polyketides',\n",
       "       'Folding, Sorting and Degradation;Poorly Characterized',\n",
       "       'Energy Metabolism;Metabolism of Other Amino Acids',\n",
       "       'Metabolism of Other Amino Acids;Nucleotide Metabolism',\n",
       "       'Genetic Information Processing;Xenobiotics Biodegradation and Metabolism',\n",
       "       'Cellular Processes and Signaling;Metabolism of Terpenoids and Polyketides',\n",
       "       'Environmental Adaptation',\n",
       "       'Enzyme Families;Glycan Biosynthesis and Metabolism',\n",
       "       'Energy Metabolism;Signal Transduction',\n",
       "       'Cellular Processes and Signaling;Metabolism of Cofactors and Vitamins',\n",
       "       'Cellular Processes and Signaling;Genetic Information Processing',\n",
       "       'Signal Transduction;unknown',\n",
       "       'Amino Acid Metabolism;Glycan Biosynthesis and Metabolism',\n",
       "       'Cellular Processes and Signaling;Metabolism;Nucleotide Metabolism',\n",
       "       'Membrane Transport;Nucleotide Metabolism',\n",
       "       'Metabolism of Cofactors and Vitamins;unknown',\n",
       "       'Amino Acid Metabolism;Lipid Metabolism',\n",
       "       'Enzyme Families;Nucleotide Metabolism',\n",
       "       'Metabolism of Cofactors and Vitamins;Transcription',\n",
       "       'Lipid Metabolism;Metabolism of Cofactors and Vitamins',\n",
       "       'Genetic Information Processing;Translation',\n",
       "       'Enzyme Families;Transcription',\n",
       "       'Biosynthesis of Other Secondary Metabolites;Metabolism of Cofactors and Vitamins',\n",
       "       'Metabolism of Cofactors and Vitamins;Metabolism of Terpenoids and Polyketides',\n",
       "       'Amino Acid Metabolism;Metabolism of Terpenoids and Polyketides',\n",
       "       'Digestive System',\n",
       "       'Nucleotide Metabolism;Xenobiotics Biodegradation and Metabolism',\n",
       "       'Cellular Processes and Signaling;Energy Metabolism;Metabolism',\n",
       "       'Genetic Information Processing;Nucleotide Metabolism', 'Cancers',\n",
       "       'Enzyme Families;Membrane Transport',\n",
       "       'Amino Acid Metabolism;Translation',\n",
       "       'Carbohydrate Metabolism;Folding, Sorting and Degradation;Poorly Characterized',\n",
       "       'Enzyme Families;Genetic Information Processing',\n",
       "       'Cellular Processes and Signaling;Energy Metabolism',\n",
       "       'Digestive System;Metabolism',\n",
       "       'Carbohydrate Metabolism;Folding, Sorting and Degradation',\n",
       "       'Glycan Biosynthesis and Metabolism;Nucleotide Metabolism',\n",
       "       'Genetic Information Processing;Glycan Biosynthesis and Metabolism',\n",
       "       'Genetic Information Processing;Metabolism;Nucleotide Metabolism',\n",
       "       'Energy Metabolism;Replication and Repair',\n",
       "       'Energy Metabolism;Poorly Characterized',\n",
       "       'Transcription;Translation',\n",
       "       'Enzyme Families;Genetic Information Processing;Glycan Biosynthesis and Metabolism',\n",
       "       'Enzyme Families;Genetic Information Processing;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism',\n",
       "       'Enzyme Families;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism',\n",
       "       'Membrane Transport;Metabolism',\n",
       "       'Enzyme Families;Genetic Information Processing;Nucleotide Metabolism',\n",
       "       'Genetic Information Processing;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism',\n",
       "       'Energy Metabolism;Genetic Information Processing'], dtype=object)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['KEGG Functional Categories'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('unknown', 512851),\n",
       " ('Carbohydrate Metabolism', 90254),\n",
       " ('Membrane Transport', 72775),\n",
       " ('Cellular Processes and Signaling', 61821),\n",
       " ('Poorly Characterized', 61400),\n",
       " ('Amino Acid Metabolism', 56274),\n",
       " ('Metabolism', 54868),\n",
       " ('Genetic Information Processing', 49299),\n",
       " ('Translation', 46951),\n",
       " ('Replication and Repair', 46526),\n",
       " ('Transcription', 36173),\n",
       " ('Nucleotide Metabolism', 34335),\n",
       " ('Energy Metabolism', 27446),\n",
       " ('Metabolism of Cofactors and Vitamins', 26134),\n",
       " ('Enzyme Families', 22267),\n",
       " ('Folding, Sorting and Degradation', 17248),\n",
       " ('Signal Transduction', 16494),\n",
       " ('Lipid Metabolism', 13502),\n",
       " ('Glycan Biosynthesis and Metabolism', 13240),\n",
       " ('Metabolism of Terpenoids and Polyketides', 8769),\n",
       " ('Metabolism of Other Amino Acids', 4782),\n",
       " ('Xenobiotics Biodegradation and Metabolism', 3118),\n",
       " ('Signaling Molecules and Interaction', 2097),\n",
       " ('Cell Motility', 1655),\n",
       " ('Infectious Diseases', 1603),\n",
       " ('Biosynthesis of Other Secondary Metabolites', 1236),\n",
       " ('Cell Growth and Death', 524),\n",
       " ('Transport and Catabolism', 279),\n",
       " ('Carbohydrate Metabolism;Metabolism', 278),\n",
       " ('Endocrine System', 149),\n",
       " ('Enzyme Families;Metabolism', 119),\n",
       " ('Cellular Processes and Signaling;Metabolism', 60),\n",
       " ('Signal Transduction;Transcription', 57),\n",
       " ('Excretory System', 56),\n",
       " ('Cancers', 52),\n",
       " ('Glycan Biosynthesis and Metabolism;Metabolism', 50),\n",
       " ('Digestive System', 49),\n",
       " ('Amino Acid Metabolism;Metabolism', 37),\n",
       " ('Metabolism;Signal Transduction', 33),\n",
       " ('Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism', 32),\n",
       " ('Amino Acid Metabolism;Cellular Processes and Signaling', 26),\n",
       " ('Lipid Metabolism;Metabolism', 24),\n",
       " ('Cellular Processes and Signaling;Poorly Characterized', 23),\n",
       " ('Carbohydrate Metabolism;Poorly Characterized', 21),\n",
       " ('Genetic Information Processing;Replication and Repair', 19),\n",
       " ('Carbohydrate Metabolism;Energy Metabolism', 18),\n",
       " ('Cell Growth and Death;Signal Transduction', 18),\n",
       " ('Carbohydrate Metabolism;Lipid Metabolism', 18),\n",
       " ('Cellular Processes and Signaling;Glycan Biosynthesis and Metabolism', 18),\n",
       " ('Replication and Repair;Transcription', 18),\n",
       " ('Carbohydrate Metabolism;Transcription', 17),\n",
       " ('Amino Acid Metabolism;Poorly Characterized', 16),\n",
       " ('Cellular Processes and Signaling;Signal Transduction', 15),\n",
       " ('Amino Acid Metabolism;Genetic Information Processing', 14),\n",
       " ('Cellular Processes and Signaling;Enzyme Families', 14),\n",
       " ('Carbohydrate Metabolism;Metabolism of Terpenoids and Polyketides', 14),\n",
       " ('Folding, Sorting and Degradation;Genetic Information Processing', 14),\n",
       " ('Metabolism;Metabolism of Cofactors and Vitamins', 13),\n",
       " ('Genetic Information Processing;Nucleotide Metabolism', 13),\n",
       " ('Genetic Information Processing;Metabolism', 12),\n",
       " ('Genetic Information Processing;Membrane Transport', 12),\n",
       " ('Carbohydrate Metabolism;Cellular Processes and Signaling', 12),\n",
       " ('Carbohydrate Metabolism;Metabolism of Cofactors and Vitamins', 12),\n",
       " ('Enzyme Families;Nucleotide Metabolism', 12),\n",
       " ('Folding, Sorting and Degradation;Translation', 11),\n",
       " ('Energy Metabolism;Metabolism', 10),\n",
       " ('Metabolism;Nucleotide Metabolism', 9),\n",
       " ('Carbohydrate Metabolism;Membrane Transport', 9),\n",
       " ('Amino Acid Metabolism;Carbohydrate Metabolism', 9),\n",
       " ('Lipid Metabolism;Xenobiotics Biodegradation and Metabolism', 9),\n",
       " ('Metabolism of Cofactors and Vitamins;Poorly Characterized', 8),\n",
       " ('Amino Acid Metabolism;Energy Metabolism', 8),\n",
       " ('Energy Metabolism;Nucleotide Metabolism', 8),\n",
       " ('Metabolism;Metabolism of Other Amino Acids', 7),\n",
       " ('Lipid Metabolism;Poorly Characterized', 7),\n",
       " ('Nucleotide Metabolism;Poorly Characterized', 7),\n",
       " ('Carbohydrate Metabolism;Nucleotide Metabolism', 7),\n",
       " ('Replication and Repair;Signal Transduction', 7),\n",
       " ('Metabolism;Xenobiotics Biodegradation and Metabolism', 7),\n",
       " ('Nucleotide Metabolism;Replication and Repair', 6),\n",
       " ('Cellular Processes and Signaling;Folding, Sorting and Degradation', 6),\n",
       " ('Genetic Information Processing;Glycan Biosynthesis and Metabolism', 6),\n",
       " ('Amino Acid Metabolism;Enzyme Families', 5),\n",
       " ('Energy Metabolism;Xenobiotics Biodegradation and Metabolism', 5),\n",
       " ('Cellular Processes and Signaling;Replication and Repair', 5),\n",
       " ('Metabolism;Poorly Characterized', 5),\n",
       " ('Folding, Sorting and Degradation;Metabolism of Cofactors and Vitamins', 5),\n",
       " ('Amino Acid Metabolism;Metabolism of Cofactors and Vitamins', 5),\n",
       " ('Metabolism;Replication and Repair', 4),\n",
       " ('Cell Growth and Death;Metabolism', 4),\n",
       " ('Enzyme Families;Replication and Repair', 4),\n",
       " ('Metabolism;Translation', 4),\n",
       " ('Membrane Transport;Signal Transduction', 4),\n",
       " ('Cellular Processes and Signaling;Metabolism of Cofactors and Vitamins', 4),\n",
       " ('Nucleotide Metabolism;Xenobiotics Biodegradation and Metabolism', 4),\n",
       " ('Enzyme Families;Genetic Information Processing', 4),\n",
       " ('Digestive System;Metabolism', 4),\n",
       " ('Cellular Processes and Signaling;Lipid Metabolism', 3),\n",
       " ('Cellular Processes and Signaling;Membrane Transport', 3),\n",
       " ('Enzyme Families;Folding, Sorting and Degradation', 3),\n",
       " ('Metabolism of Other Amino Acids;Transcription', 3),\n",
       " ('Metabolism;Metabolism of Terpenoids and Polyketides', 3),\n",
       " ('Carbohydrate Metabolism;Signal Transduction', 3),\n",
       " ('Poorly Characterized;Xenobiotics Biodegradation and Metabolism', 3),\n",
       " ('Cellular Processes and Signaling;Genetic Information Processing', 3),\n",
       " ('Metabolism of Cofactors and Vitamins;Transcription', 3),\n",
       " ('Energy Metabolism;Poorly Characterized', 3),\n",
       " ('Enzyme Families;Genetic Information Processing;Glycan Biosynthesis and Metabolism',\n",
       "  3),\n",
       " ('Carbohydrate Metabolism;Enzyme Families', 2),\n",
       " ('Enzyme Families;Signal Transduction', 2),\n",
       " ('Poorly Characterized;Translation', 2),\n",
       " ('Genetic Information Processing;Transcription', 2),\n",
       " ('Cellular Processes and Signaling;Transcription', 2),\n",
       " ('Cellular Processes and Signaling;Metabolism of Other Amino Acids', 2),\n",
       " ('Membrane Transport;Replication and Repair', 2),\n",
       " ('Enzyme Families;Metabolism;unknown', 2),\n",
       " ('Immune System Diseases', 2),\n",
       " ('Enzyme Families;Transport and Catabolism', 2),\n",
       " ('Biosynthesis of Other Secondary Metabolites;Lipid Metabolism', 2),\n",
       " ('Metabolism;unknown', 2),\n",
       " ('Carbohydrate Metabolism;Replication and Repair', 2),\n",
       " ('Metabolism of Cofactors and Vitamins;Xenobiotics Biodegradation and Metabolism',\n",
       "  2),\n",
       " ('Folding, Sorting and Degradation;Metabolism', 2),\n",
       " ('Amino Acid Metabolism;Xenobiotics Biodegradation and Metabolism', 2),\n",
       " ('Metabolism of Terpenoids and Polyketides;Translation', 2),\n",
       " ('Biosynthesis of Other Secondary Metabolites;Xenobiotics Biodegradation and Metabolism',\n",
       "  2),\n",
       " ('Environmental Adaptation', 2),\n",
       " ('Enzyme Families;Glycan Biosynthesis and Metabolism', 2),\n",
       " ('Signal Transduction;unknown', 2),\n",
       " ('Membrane Transport;Nucleotide Metabolism', 2),\n",
       " ('Metabolism of Cofactors and Vitamins;Metabolism of Terpenoids and Polyketides',\n",
       "  2),\n",
       " ('Amino Acid Metabolism;Translation', 2),\n",
       " ('Energy Metabolism;Replication and Repair', 2),\n",
       " ('Cellular Processes and Signaling;Signaling Molecules and Interaction', 1),\n",
       " ('Membrane Transport;Signaling Molecules and Interaction', 1),\n",
       " ('Metabolism;Signaling Molecules and Interaction', 1),\n",
       " ('Folding, Sorting and Degradation;Infectious Diseases', 1),\n",
       " ('Infectious Diseases;Poorly Characterized', 1),\n",
       " ('Amino Acid Metabolism;Infectious Diseases', 1),\n",
       " ('Cellular Processes and Signaling;Enzyme Families;Metabolism', 1),\n",
       " ('Infectious Diseases;Metabolism', 1),\n",
       " ('Enzyme Families;Metabolism;Signal Transduction', 1),\n",
       " ('Glycan Biosynthesis and Metabolism;Metabolism of Terpenoids and Polyketides',\n",
       "  1),\n",
       " ('Membrane Transport;Poorly Characterized', 1),\n",
       " ('Energy Metabolism;Folding, Sorting and Degradation;Metabolism', 1),\n",
       " ('Cellular Processes and Signaling;Infectious Diseases', 1),\n",
       " ('Glycan Biosynthesis and Metabolism;Lipid Metabolism', 1),\n",
       " ('Lipid Metabolism;Metabolism of Terpenoids and Polyketides', 1),\n",
       " ('Enzyme Families;unknown', 1),\n",
       " ('Cell Growth and Death;Enzyme Families', 1),\n",
       " ('Cellular Processes and Signaling;Nucleotide Metabolism', 1),\n",
       " ('Metabolism of Other Amino Acids;Xenobiotics Biodegradation and Metabolism',\n",
       "  1),\n",
       " ('Cellular Processes and Signaling;unknown', 1),\n",
       " ('Carbohydrate Metabolism;Signaling Molecules and Interaction', 1),\n",
       " ('Lipid Metabolism;Replication and Repair', 1),\n",
       " ('Enzyme Families;Signaling Molecules and Interaction', 1),\n",
       " ('Folding, Sorting and Degradation;Membrane Transport', 1),\n",
       " ('Folding, Sorting and Degradation;Genetic Information Processing;Translation',\n",
       "  1),\n",
       " ('Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism;Metabolism', 1),\n",
       " ('Carbohydrate Metabolism;Transport and Catabolism', 1),\n",
       " ('Cell Motility;Membrane Transport', 1),\n",
       " ('Metabolism;Transcription', 1),\n",
       " ('Lipid Metabolism;Transcription', 1),\n",
       " ('Poorly Characterized;Signal Transduction', 1),\n",
       " ('Carbohydrate Metabolism;Cellular Processes and Signaling;Metabolism', 1),\n",
       " ('Cell Motility;Poorly Characterized', 1),\n",
       " ('Amino Acid Metabolism;Transcription', 1),\n",
       " ('Membrane Transport;Transcription', 1),\n",
       " ('Metabolism of Cofactors and Vitamins;Metabolism of Other Amino Acids', 1),\n",
       " ('Carbohydrate Metabolism;Metabolism;Metabolism of Terpenoids and Polyketides',\n",
       "  1),\n",
       " ('Folding, Sorting and Degradation;Poorly Characterized', 1),\n",
       " ('Energy Metabolism;Metabolism of Other Amino Acids', 1),\n",
       " ('Metabolism of Other Amino Acids;Nucleotide Metabolism', 1),\n",
       " ('Genetic Information Processing;Xenobiotics Biodegradation and Metabolism',\n",
       "  1),\n",
       " ('Cellular Processes and Signaling;Metabolism of Terpenoids and Polyketides',\n",
       "  1),\n",
       " ('Energy Metabolism;Signal Transduction', 1),\n",
       " ('Amino Acid Metabolism;Glycan Biosynthesis and Metabolism', 1),\n",
       " ('Cellular Processes and Signaling;Metabolism;Nucleotide Metabolism', 1),\n",
       " ('Metabolism of Cofactors and Vitamins;unknown', 1),\n",
       " ('Amino Acid Metabolism;Lipid Metabolism', 1),\n",
       " ('Lipid Metabolism;Metabolism of Cofactors and Vitamins', 1),\n",
       " ('Genetic Information Processing;Translation', 1),\n",
       " ('Enzyme Families;Transcription', 1),\n",
       " ('Biosynthesis of Other Secondary Metabolites;Metabolism of Cofactors and Vitamins',\n",
       "  1),\n",
       " ('Amino Acid Metabolism;Metabolism of Terpenoids and Polyketides', 1),\n",
       " ('Cellular Processes and Signaling;Energy Metabolism;Metabolism', 1),\n",
       " ('Enzyme Families;Membrane Transport', 1),\n",
       " ('Carbohydrate Metabolism;Folding, Sorting and Degradation;Poorly Characterized',\n",
       "  1),\n",
       " ('Cellular Processes and Signaling;Energy Metabolism', 1),\n",
       " ('Carbohydrate Metabolism;Folding, Sorting and Degradation', 1),\n",
       " ('Glycan Biosynthesis and Metabolism;Nucleotide Metabolism', 1),\n",
       " ('Genetic Information Processing;Metabolism;Nucleotide Metabolism', 1),\n",
       " ('Transcription;Translation', 1),\n",
       " ('Enzyme Families;Genetic Information Processing;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism',\n",
       "  1),\n",
       " ('Enzyme Families;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism',\n",
       "  1),\n",
       " ('Membrane Transport;Metabolism', 1),\n",
       " ('Enzyme Families;Genetic Information Processing;Nucleotide Metabolism', 1),\n",
       " ('Genetic Information Processing;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism',\n",
       "  1),\n",
       " ('Energy Metabolism;Genetic Information Processing', 1)]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(df['KEGG Functional Categories']).most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg = []\n",
    "\n",
    "for category in df['KEGG Functional Categories']:\n",
    "    if category not in kegg:\n",
    "        kegg.append(category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg = {}\n",
    "\n",
    "for category in df['KEGG Functional Categories']:\n",
    "    if category not in kegg:\n",
    "        kegg[category] = 1\n",
    "    else:\n",
    "        kegg[category] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('unknown', 512851),\n",
       " ('Carbohydrate Metabolism', 90254),\n",
       " ('Membrane Transport', 72775),\n",
       " ('Cellular Processes and Signaling', 61821),\n",
       " ('Poorly Characterized', 61400),\n",
       " ('Amino Acid Metabolism', 56274),\n",
       " ('Metabolism', 54868),\n",
       " ('Genetic Information Processing', 49299),\n",
       " ('Translation', 46951),\n",
       " ('Replication and Repair', 46526),\n",
       " ('Transcription', 36173),\n",
       " ('Nucleotide Metabolism', 34335),\n",
       " ('Energy Metabolism', 27446),\n",
       " ('Metabolism of Cofactors and Vitamins', 26134),\n",
       " ('Enzyme Families', 22267),\n",
       " ('Folding, Sorting and Degradation', 17248),\n",
       " ('Signal Transduction', 16494),\n",
       " ('Lipid Metabolism', 13502),\n",
       " ('Glycan Biosynthesis and Metabolism', 13240),\n",
       " ('Metabolism of Terpenoids and Polyketides', 8769),\n",
       " ('Metabolism of Other Amino Acids', 4782),\n",
       " ('Xenobiotics Biodegradation and Metabolism', 3118),\n",
       " ('Signaling Molecules and Interaction', 2097),\n",
       " ('Cell Motility', 1655),\n",
       " ('Infectious Diseases', 1603),\n",
       " ('Biosynthesis of Other Secondary Metabolites', 1236),\n",
       " ('Cell Growth and Death', 524),\n",
       " ('Transport and Catabolism', 279),\n",
       " ('Carbohydrate Metabolism;Metabolism', 278),\n",
       " ('Endocrine System', 149),\n",
       " ('Enzyme Families;Metabolism', 119),\n",
       " ('Cellular Processes and Signaling;Metabolism', 60),\n",
       " ('Signal Transduction;Transcription', 57),\n",
       " ('Excretory System', 56),\n",
       " ('Cancers', 52),\n",
       " ('Glycan Biosynthesis and Metabolism;Metabolism', 50),\n",
       " ('Digestive System', 49),\n",
       " ('Amino Acid Metabolism;Metabolism', 37),\n",
       " ('Metabolism;Signal Transduction', 33),\n",
       " ('Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism', 32),\n",
       " ('Amino Acid Metabolism;Cellular Processes and Signaling', 26),\n",
       " ('Lipid Metabolism;Metabolism', 24),\n",
       " ('Cellular Processes and Signaling;Poorly Characterized', 23),\n",
       " ('Carbohydrate Metabolism;Poorly Characterized', 21),\n",
       " ('Genetic Information Processing;Replication and Repair', 19),\n",
       " ('Carbohydrate Metabolism;Energy Metabolism', 18),\n",
       " ('Cell Growth and Death;Signal Transduction', 18),\n",
       " ('Carbohydrate Metabolism;Lipid Metabolism', 18),\n",
       " ('Cellular Processes and Signaling;Glycan Biosynthesis and Metabolism', 18),\n",
       " ('Replication and Repair;Transcription', 18),\n",
       " ('Carbohydrate Metabolism;Transcription', 17),\n",
       " ('Amino Acid Metabolism;Poorly Characterized', 16),\n",
       " ('Cellular Processes and Signaling;Signal Transduction', 15),\n",
       " ('Amino Acid Metabolism;Genetic Information Processing', 14),\n",
       " ('Cellular Processes and Signaling;Enzyme Families', 14),\n",
       " ('Carbohydrate Metabolism;Metabolism of Terpenoids and Polyketides', 14),\n",
       " ('Folding, Sorting and Degradation;Genetic Information Processing', 14),\n",
       " ('Metabolism;Metabolism of Cofactors and Vitamins', 13),\n",
       " ('Genetic Information Processing;Nucleotide Metabolism', 13),\n",
       " ('Genetic Information Processing;Metabolism', 12),\n",
       " ('Genetic Information Processing;Membrane Transport', 12),\n",
       " ('Carbohydrate Metabolism;Cellular Processes and Signaling', 12),\n",
       " ('Carbohydrate Metabolism;Metabolism of Cofactors and Vitamins', 12),\n",
       " ('Enzyme Families;Nucleotide Metabolism', 12),\n",
       " ('Folding, Sorting and Degradation;Translation', 11),\n",
       " ('Energy Metabolism;Metabolism', 10),\n",
       " ('Metabolism;Nucleotide Metabolism', 9),\n",
       " ('Carbohydrate Metabolism;Membrane Transport', 9),\n",
       " ('Amino Acid Metabolism;Carbohydrate Metabolism', 9),\n",
       " ('Lipid Metabolism;Xenobiotics Biodegradation and Metabolism', 9),\n",
       " ('Metabolism of Cofactors and Vitamins;Poorly Characterized', 8),\n",
       " ('Amino Acid Metabolism;Energy Metabolism', 8),\n",
       " ('Energy Metabolism;Nucleotide Metabolism', 8),\n",
       " ('Metabolism;Metabolism of Other Amino Acids', 7),\n",
       " ('Lipid Metabolism;Poorly Characterized', 7),\n",
       " ('Nucleotide Metabolism;Poorly Characterized', 7),\n",
       " ('Carbohydrate Metabolism;Nucleotide Metabolism', 7),\n",
       " ('Replication and Repair;Signal Transduction', 7),\n",
       " ('Metabolism;Xenobiotics Biodegradation and Metabolism', 7),\n",
       " ('Nucleotide Metabolism;Replication and Repair', 6),\n",
       " ('Cellular Processes and Signaling;Folding, Sorting and Degradation', 6),\n",
       " ('Genetic Information Processing;Glycan Biosynthesis and Metabolism', 6),\n",
       " ('Amino Acid Metabolism;Enzyme Families', 5),\n",
       " ('Energy Metabolism;Xenobiotics Biodegradation and Metabolism', 5),\n",
       " ('Cellular Processes and Signaling;Replication and Repair', 5),\n",
       " ('Metabolism;Poorly Characterized', 5),\n",
       " ('Folding, Sorting and Degradation;Metabolism of Cofactors and Vitamins', 5),\n",
       " ('Amino Acid Metabolism;Metabolism of Cofactors and Vitamins', 5),\n",
       " ('Metabolism;Replication and Repair', 4),\n",
       " ('Cell Growth and Death;Metabolism', 4),\n",
       " ('Enzyme Families;Replication and Repair', 4),\n",
       " ('Metabolism;Translation', 4),\n",
       " ('Membrane Transport;Signal Transduction', 4),\n",
       " ('Cellular Processes and Signaling;Metabolism of Cofactors and Vitamins', 4),\n",
       " ('Nucleotide Metabolism;Xenobiotics Biodegradation and Metabolism', 4),\n",
       " ('Enzyme Families;Genetic Information Processing', 4),\n",
       " ('Digestive System;Metabolism', 4),\n",
       " ('Cellular Processes and Signaling;Lipid Metabolism', 3),\n",
       " ('Cellular Processes and Signaling;Membrane Transport', 3),\n",
       " ('Enzyme Families;Folding, Sorting and Degradation', 3),\n",
       " ('Metabolism of Other Amino Acids;Transcription', 3),\n",
       " ('Metabolism;Metabolism of Terpenoids and Polyketides', 3),\n",
       " ('Carbohydrate Metabolism;Signal Transduction', 3),\n",
       " ('Poorly Characterized;Xenobiotics Biodegradation and Metabolism', 3),\n",
       " ('Cellular Processes and Signaling;Genetic Information Processing', 3),\n",
       " ('Metabolism of Cofactors and Vitamins;Transcription', 3),\n",
       " ('Energy Metabolism;Poorly Characterized', 3),\n",
       " ('Enzyme Families;Genetic Information Processing;Glycan Biosynthesis and Metabolism',\n",
       "  3),\n",
       " ('Carbohydrate Metabolism;Enzyme Families', 2),\n",
       " ('Enzyme Families;Signal Transduction', 2),\n",
       " ('Poorly Characterized;Translation', 2),\n",
       " ('Genetic Information Processing;Transcription', 2),\n",
       " ('Cellular Processes and Signaling;Transcription', 2),\n",
       " ('Cellular Processes and Signaling;Metabolism of Other Amino Acids', 2),\n",
       " ('Membrane Transport;Replication and Repair', 2),\n",
       " ('Enzyme Families;Metabolism;unknown', 2),\n",
       " ('Immune System Diseases', 2),\n",
       " ('Enzyme Families;Transport and Catabolism', 2),\n",
       " ('Biosynthesis of Other Secondary Metabolites;Lipid Metabolism', 2),\n",
       " ('Metabolism;unknown', 2),\n",
       " ('Carbohydrate Metabolism;Replication and Repair', 2),\n",
       " ('Metabolism of Cofactors and Vitamins;Xenobiotics Biodegradation and Metabolism',\n",
       "  2),\n",
       " ('Folding, Sorting and Degradation;Metabolism', 2),\n",
       " ('Amino Acid Metabolism;Xenobiotics Biodegradation and Metabolism', 2),\n",
       " ('Metabolism of Terpenoids and Polyketides;Translation', 2),\n",
       " ('Biosynthesis of Other Secondary Metabolites;Xenobiotics Biodegradation and Metabolism',\n",
       "  2),\n",
       " ('Environmental Adaptation', 2),\n",
       " ('Enzyme Families;Glycan Biosynthesis and Metabolism', 2),\n",
       " ('Signal Transduction;unknown', 2),\n",
       " ('Membrane Transport;Nucleotide Metabolism', 2),\n",
       " ('Metabolism of Cofactors and Vitamins;Metabolism of Terpenoids and Polyketides',\n",
       "  2),\n",
       " ('Amino Acid Metabolism;Translation', 2),\n",
       " ('Energy Metabolism;Replication and Repair', 2),\n",
       " ('Cellular Processes and Signaling;Signaling Molecules and Interaction', 1),\n",
       " ('Membrane Transport;Signaling Molecules and Interaction', 1),\n",
       " ('Metabolism;Signaling Molecules and Interaction', 1),\n",
       " ('Folding, Sorting and Degradation;Infectious Diseases', 1),\n",
       " ('Infectious Diseases;Poorly Characterized', 1),\n",
       " ('Amino Acid Metabolism;Infectious Diseases', 1),\n",
       " ('Cellular Processes and Signaling;Enzyme Families;Metabolism', 1),\n",
       " ('Infectious Diseases;Metabolism', 1),\n",
       " ('Enzyme Families;Metabolism;Signal Transduction', 1),\n",
       " ('Glycan Biosynthesis and Metabolism;Metabolism of Terpenoids and Polyketides',\n",
       "  1),\n",
       " ('Membrane Transport;Poorly Characterized', 1),\n",
       " ('Energy Metabolism;Folding, Sorting and Degradation;Metabolism', 1),\n",
       " ('Cellular Processes and Signaling;Infectious Diseases', 1),\n",
       " ('Glycan Biosynthesis and Metabolism;Lipid Metabolism', 1),\n",
       " ('Lipid Metabolism;Metabolism of Terpenoids and Polyketides', 1),\n",
       " ('Enzyme Families;unknown', 1),\n",
       " ('Cell Growth and Death;Enzyme Families', 1),\n",
       " ('Cellular Processes and Signaling;Nucleotide Metabolism', 1),\n",
       " ('Metabolism of Other Amino Acids;Xenobiotics Biodegradation and Metabolism',\n",
       "  1),\n",
       " ('Cellular Processes and Signaling;unknown', 1),\n",
       " ('Carbohydrate Metabolism;Signaling Molecules and Interaction', 1),\n",
       " ('Lipid Metabolism;Replication and Repair', 1),\n",
       " ('Enzyme Families;Signaling Molecules and Interaction', 1),\n",
       " ('Folding, Sorting and Degradation;Membrane Transport', 1),\n",
       " ('Folding, Sorting and Degradation;Genetic Information Processing;Translation',\n",
       "  1),\n",
       " ('Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism;Metabolism', 1),\n",
       " ('Carbohydrate Metabolism;Transport and Catabolism', 1),\n",
       " ('Cell Motility;Membrane Transport', 1),\n",
       " ('Metabolism;Transcription', 1),\n",
       " ('Lipid Metabolism;Transcription', 1),\n",
       " ('Poorly Characterized;Signal Transduction', 1),\n",
       " ('Carbohydrate Metabolism;Cellular Processes and Signaling;Metabolism', 1),\n",
       " ('Cell Motility;Poorly Characterized', 1),\n",
       " ('Amino Acid Metabolism;Transcription', 1),\n",
       " ('Membrane Transport;Transcription', 1),\n",
       " ('Metabolism of Cofactors and Vitamins;Metabolism of Other Amino Acids', 1),\n",
       " ('Carbohydrate Metabolism;Metabolism;Metabolism of Terpenoids and Polyketides',\n",
       "  1),\n",
       " ('Folding, Sorting and Degradation;Poorly Characterized', 1),\n",
       " ('Energy Metabolism;Metabolism of Other Amino Acids', 1),\n",
       " ('Metabolism of Other Amino Acids;Nucleotide Metabolism', 1),\n",
       " ('Genetic Information Processing;Xenobiotics Biodegradation and Metabolism',\n",
       "  1),\n",
       " ('Cellular Processes and Signaling;Metabolism of Terpenoids and Polyketides',\n",
       "  1),\n",
       " ('Energy Metabolism;Signal Transduction', 1),\n",
       " ('Amino Acid Metabolism;Glycan Biosynthesis and Metabolism', 1),\n",
       " ('Cellular Processes and Signaling;Metabolism;Nucleotide Metabolism', 1),\n",
       " ('Metabolism of Cofactors and Vitamins;unknown', 1),\n",
       " ('Amino Acid Metabolism;Lipid Metabolism', 1),\n",
       " ('Lipid Metabolism;Metabolism of Cofactors and Vitamins', 1),\n",
       " ('Genetic Information Processing;Translation', 1),\n",
       " ('Enzyme Families;Transcription', 1),\n",
       " ('Biosynthesis of Other Secondary Metabolites;Metabolism of Cofactors and Vitamins',\n",
       "  1),\n",
       " ('Amino Acid Metabolism;Metabolism of Terpenoids and Polyketides', 1),\n",
       " ('Cellular Processes and Signaling;Energy Metabolism;Metabolism', 1),\n",
       " ('Enzyme Families;Membrane Transport', 1),\n",
       " ('Carbohydrate Metabolism;Folding, Sorting and Degradation;Poorly Characterized',\n",
       "  1),\n",
       " ('Cellular Processes and Signaling;Energy Metabolism', 1),\n",
       " ('Carbohydrate Metabolism;Folding, Sorting and Degradation', 1),\n",
       " ('Glycan Biosynthesis and Metabolism;Nucleotide Metabolism', 1),\n",
       " ('Genetic Information Processing;Metabolism;Nucleotide Metabolism', 1),\n",
       " ('Transcription;Translation', 1),\n",
       " ('Enzyme Families;Genetic Information Processing;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism',\n",
       "  1),\n",
       " ('Enzyme Families;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism',\n",
       "  1),\n",
       " ('Membrane Transport;Metabolism', 1),\n",
       " ('Enzyme Families;Genetic Information Processing;Nucleotide Metabolism', 1),\n",
       " ('Genetic Information Processing;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism',\n",
       "  1),\n",
       " ('Energy Metabolism;Genetic Information Processing', 1)]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(kegg.items(), key=lambda x: x[1], reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'unknown': 512851,\n",
       " 'Carbohydrate Metabolism': 90254,\n",
       " 'Membrane Transport': 72775,\n",
       " 'Cellular Processes and Signaling': 61821,\n",
       " 'Poorly Characterized': 61400,\n",
       " 'Amino Acid Metabolism': 56274,\n",
       " 'Metabolism': 54868,\n",
       " 'Genetic Information Processing': 49299,\n",
       " 'Translation': 46951,\n",
       " 'Replication and Repair': 46526,\n",
       " 'Transcription': 36173,\n",
       " 'Nucleotide Metabolism': 34335,\n",
       " 'Energy Metabolism': 27446,\n",
       " 'Metabolism of Cofactors and Vitamins': 26134,\n",
       " 'Enzyme Families': 22267,\n",
       " 'Folding, Sorting and Degradation': 17248,\n",
       " 'Signal Transduction': 16494,\n",
       " 'Lipid Metabolism': 13502,\n",
       " 'Glycan Biosynthesis and Metabolism': 13240,\n",
       " 'Metabolism of Terpenoids and Polyketides': 8769,\n",
       " 'Metabolism of Other Amino Acids': 4782,\n",
       " 'Xenobiotics Biodegradation and Metabolism': 3118,\n",
       " 'Signaling Molecules and Interaction': 2097,\n",
       " 'Cell Motility': 1655,\n",
       " 'Infectious Diseases': 1603,\n",
       " 'Biosynthesis of Other Secondary Metabolites': 1236,\n",
       " 'Cell Growth and Death': 524,\n",
       " 'Transport and Catabolism': 279,\n",
       " 'Carbohydrate Metabolism;Metabolism': 278,\n",
       " 'Endocrine System': 149,\n",
       " 'Enzyme Families;Metabolism': 119,\n",
       " 'Cellular Processes and Signaling;Metabolism': 60,\n",
       " 'Signal Transduction;Transcription': 57,\n",
       " 'Excretory System': 56,\n",
       " 'Cancers': 52,\n",
       " 'Glycan Biosynthesis and Metabolism;Metabolism': 50,\n",
       " 'Digestive System': 49,\n",
       " 'Amino Acid Metabolism;Metabolism': 37,\n",
       " 'Metabolism;Signal Transduction': 33,\n",
       " 'Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism': 32,\n",
       " 'Amino Acid Metabolism;Cellular Processes and Signaling': 26,\n",
       " 'Lipid Metabolism;Metabolism': 24,\n",
       " 'Cellular Processes and Signaling;Poorly Characterized': 23,\n",
       " 'Carbohydrate Metabolism;Poorly Characterized': 21,\n",
       " 'Genetic Information Processing;Replication and Repair': 19,\n",
       " 'Carbohydrate Metabolism;Energy Metabolism': 18,\n",
       " 'Cell Growth and Death;Signal Transduction': 18,\n",
       " 'Carbohydrate Metabolism;Lipid Metabolism': 18,\n",
       " 'Cellular Processes and Signaling;Glycan Biosynthesis and Metabolism': 18,\n",
       " 'Replication and Repair;Transcription': 18,\n",
       " 'Carbohydrate Metabolism;Transcription': 17,\n",
       " 'Amino Acid Metabolism;Poorly Characterized': 16,\n",
       " 'Cellular Processes and Signaling;Signal Transduction': 15,\n",
       " 'Amino Acid Metabolism;Genetic Information Processing': 14,\n",
       " 'Cellular Processes and Signaling;Enzyme Families': 14,\n",
       " 'Carbohydrate Metabolism;Metabolism of Terpenoids and Polyketides': 14,\n",
       " 'Folding, Sorting and Degradation;Genetic Information Processing': 14,\n",
       " 'Metabolism;Metabolism of Cofactors and Vitamins': 13,\n",
       " 'Genetic Information Processing;Nucleotide Metabolism': 13,\n",
       " 'Genetic Information Processing;Metabolism': 12,\n",
       " 'Genetic Information Processing;Membrane Transport': 12,\n",
       " 'Carbohydrate Metabolism;Cellular Processes and Signaling': 12,\n",
       " 'Carbohydrate Metabolism;Metabolism of Cofactors and Vitamins': 12,\n",
       " 'Enzyme Families;Nucleotide Metabolism': 12,\n",
       " 'Folding, Sorting and Degradation;Translation': 11,\n",
       " 'Energy Metabolism;Metabolism': 10,\n",
       " 'Metabolism;Nucleotide Metabolism': 9,\n",
       " 'Carbohydrate Metabolism;Membrane Transport': 9,\n",
       " 'Amino Acid Metabolism;Carbohydrate Metabolism': 9,\n",
       " 'Lipid Metabolism;Xenobiotics Biodegradation and Metabolism': 9,\n",
       " 'Metabolism of Cofactors and Vitamins;Poorly Characterized': 8,\n",
       " 'Amino Acid Metabolism;Energy Metabolism': 8,\n",
       " 'Energy Metabolism;Nucleotide Metabolism': 8,\n",
       " 'Metabolism;Metabolism of Other Amino Acids': 7,\n",
       " 'Lipid Metabolism;Poorly Characterized': 7,\n",
       " 'Nucleotide Metabolism;Poorly Characterized': 7,\n",
       " 'Carbohydrate Metabolism;Nucleotide Metabolism': 7,\n",
       " 'Replication and Repair;Signal Transduction': 7,\n",
       " 'Metabolism;Xenobiotics Biodegradation and Metabolism': 7,\n",
       " 'Nucleotide Metabolism;Replication and Repair': 6,\n",
       " 'Cellular Processes and Signaling;Folding, Sorting and Degradation': 6,\n",
       " 'Genetic Information Processing;Glycan Biosynthesis and Metabolism': 6,\n",
       " 'Amino Acid Metabolism;Enzyme Families': 5,\n",
       " 'Energy Metabolism;Xenobiotics Biodegradation and Metabolism': 5,\n",
       " 'Cellular Processes and Signaling;Replication and Repair': 5,\n",
       " 'Metabolism;Poorly Characterized': 5,\n",
       " 'Folding, Sorting and Degradation;Metabolism of Cofactors and Vitamins': 5,\n",
       " 'Amino Acid Metabolism;Metabolism of Cofactors and Vitamins': 5,\n",
       " 'Metabolism;Replication and Repair': 4,\n",
       " 'Cell Growth and Death;Metabolism': 4,\n",
       " 'Enzyme Families;Replication and Repair': 4,\n",
       " 'Metabolism;Translation': 4,\n",
       " 'Membrane Transport;Signal Transduction': 4,\n",
       " 'Cellular Processes and Signaling;Metabolism of Cofactors and Vitamins': 4,\n",
       " 'Nucleotide Metabolism;Xenobiotics Biodegradation and Metabolism': 4,\n",
       " 'Enzyme Families;Genetic Information Processing': 4,\n",
       " 'Digestive System;Metabolism': 4,\n",
       " 'Cellular Processes and Signaling;Lipid Metabolism': 3,\n",
       " 'Cellular Processes and Signaling;Membrane Transport': 3,\n",
       " 'Enzyme Families;Folding, Sorting and Degradation': 3,\n",
       " 'Metabolism of Other Amino Acids;Transcription': 3,\n",
       " 'Metabolism;Metabolism of Terpenoids and Polyketides': 3,\n",
       " 'Carbohydrate Metabolism;Signal Transduction': 3,\n",
       " 'Poorly Characterized;Xenobiotics Biodegradation and Metabolism': 3,\n",
       " 'Cellular Processes and Signaling;Genetic Information Processing': 3,\n",
       " 'Metabolism of Cofactors and Vitamins;Transcription': 3,\n",
       " 'Energy Metabolism;Poorly Characterized': 3,\n",
       " 'Enzyme Families;Genetic Information Processing;Glycan Biosynthesis and Metabolism': 3,\n",
       " 'Carbohydrate Metabolism;Enzyme Families': 2,\n",
       " 'Enzyme Families;Signal Transduction': 2,\n",
       " 'Poorly Characterized;Translation': 2,\n",
       " 'Genetic Information Processing;Transcription': 2,\n",
       " 'Cellular Processes and Signaling;Transcription': 2,\n",
       " 'Cellular Processes and Signaling;Metabolism of Other Amino Acids': 2,\n",
       " 'Membrane Transport;Replication and Repair': 2,\n",
       " 'Enzyme Families;Metabolism;unknown': 2,\n",
       " 'Immune System Diseases': 2,\n",
       " 'Enzyme Families;Transport and Catabolism': 2,\n",
       " 'Biosynthesis of Other Secondary Metabolites;Lipid Metabolism': 2,\n",
       " 'Metabolism;unknown': 2,\n",
       " 'Carbohydrate Metabolism;Replication and Repair': 2,\n",
       " 'Metabolism of Cofactors and Vitamins;Xenobiotics Biodegradation and Metabolism': 2,\n",
       " 'Folding, Sorting and Degradation;Metabolism': 2,\n",
       " 'Amino Acid Metabolism;Xenobiotics Biodegradation and Metabolism': 2,\n",
       " 'Metabolism of Terpenoids and Polyketides;Translation': 2,\n",
       " 'Biosynthesis of Other Secondary Metabolites;Xenobiotics Biodegradation and Metabolism': 2,\n",
       " 'Environmental Adaptation': 2,\n",
       " 'Enzyme Families;Glycan Biosynthesis and Metabolism': 2,\n",
       " 'Signal Transduction;unknown': 2,\n",
       " 'Membrane Transport;Nucleotide Metabolism': 2,\n",
       " 'Metabolism of Cofactors and Vitamins;Metabolism of Terpenoids and Polyketides': 2,\n",
       " 'Amino Acid Metabolism;Translation': 2,\n",
       " 'Energy Metabolism;Replication and Repair': 2,\n",
       " 'Cellular Processes and Signaling;Signaling Molecules and Interaction': 1,\n",
       " 'Membrane Transport;Signaling Molecules and Interaction': 1,\n",
       " 'Metabolism;Signaling Molecules and Interaction': 1,\n",
       " 'Folding, Sorting and Degradation;Infectious Diseases': 1,\n",
       " 'Infectious Diseases;Poorly Characterized': 1,\n",
       " 'Amino Acid Metabolism;Infectious Diseases': 1,\n",
       " 'Cellular Processes and Signaling;Enzyme Families;Metabolism': 1,\n",
       " 'Infectious Diseases;Metabolism': 1,\n",
       " 'Enzyme Families;Metabolism;Signal Transduction': 1,\n",
       " 'Glycan Biosynthesis and Metabolism;Metabolism of Terpenoids and Polyketides': 1,\n",
       " 'Membrane Transport;Poorly Characterized': 1,\n",
       " 'Energy Metabolism;Folding, Sorting and Degradation;Metabolism': 1,\n",
       " 'Cellular Processes and Signaling;Infectious Diseases': 1,\n",
       " 'Glycan Biosynthesis and Metabolism;Lipid Metabolism': 1,\n",
       " 'Lipid Metabolism;Metabolism of Terpenoids and Polyketides': 1,\n",
       " 'Enzyme Families;unknown': 1,\n",
       " 'Cell Growth and Death;Enzyme Families': 1,\n",
       " 'Cellular Processes and Signaling;Nucleotide Metabolism': 1,\n",
       " 'Metabolism of Other Amino Acids;Xenobiotics Biodegradation and Metabolism': 1,\n",
       " 'Cellular Processes and Signaling;unknown': 1,\n",
       " 'Carbohydrate Metabolism;Signaling Molecules and Interaction': 1,\n",
       " 'Lipid Metabolism;Replication and Repair': 1,\n",
       " 'Enzyme Families;Signaling Molecules and Interaction': 1,\n",
       " 'Folding, Sorting and Degradation;Membrane Transport': 1,\n",
       " 'Folding, Sorting and Degradation;Genetic Information Processing;Translation': 1,\n",
       " 'Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism;Metabolism': 1,\n",
       " 'Carbohydrate Metabolism;Transport and Catabolism': 1,\n",
       " 'Cell Motility;Membrane Transport': 1,\n",
       " 'Metabolism;Transcription': 1,\n",
       " 'Lipid Metabolism;Transcription': 1,\n",
       " 'Poorly Characterized;Signal Transduction': 1,\n",
       " 'Carbohydrate Metabolism;Cellular Processes and Signaling;Metabolism': 1,\n",
       " 'Cell Motility;Poorly Characterized': 1,\n",
       " 'Amino Acid Metabolism;Transcription': 1,\n",
       " 'Membrane Transport;Transcription': 1,\n",
       " 'Metabolism of Cofactors and Vitamins;Metabolism of Other Amino Acids': 1,\n",
       " 'Carbohydrate Metabolism;Metabolism;Metabolism of Terpenoids and Polyketides': 1,\n",
       " 'Folding, Sorting and Degradation;Poorly Characterized': 1,\n",
       " 'Energy Metabolism;Metabolism of Other Amino Acids': 1,\n",
       " 'Metabolism of Other Amino Acids;Nucleotide Metabolism': 1,\n",
       " 'Genetic Information Processing;Xenobiotics Biodegradation and Metabolism': 1,\n",
       " 'Cellular Processes and Signaling;Metabolism of Terpenoids and Polyketides': 1,\n",
       " 'Energy Metabolism;Signal Transduction': 1,\n",
       " 'Amino Acid Metabolism;Glycan Biosynthesis and Metabolism': 1,\n",
       " 'Cellular Processes and Signaling;Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Metabolism of Cofactors and Vitamins;unknown': 1,\n",
       " 'Amino Acid Metabolism;Lipid Metabolism': 1,\n",
       " 'Lipid Metabolism;Metabolism of Cofactors and Vitamins': 1,\n",
       " 'Genetic Information Processing;Translation': 1,\n",
       " 'Enzyme Families;Transcription': 1,\n",
       " 'Biosynthesis of Other Secondary Metabolites;Metabolism of Cofactors and Vitamins': 1,\n",
       " 'Amino Acid Metabolism;Metabolism of Terpenoids and Polyketides': 1,\n",
       " 'Cellular Processes and Signaling;Energy Metabolism;Metabolism': 1,\n",
       " 'Enzyme Families;Membrane Transport': 1,\n",
       " 'Carbohydrate Metabolism;Folding, Sorting and Degradation;Poorly Characterized': 1,\n",
       " 'Cellular Processes and Signaling;Energy Metabolism': 1,\n",
       " 'Carbohydrate Metabolism;Folding, Sorting and Degradation': 1,\n",
       " 'Glycan Biosynthesis and Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Genetic Information Processing;Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Transcription;Translation': 1,\n",
       " 'Enzyme Families;Genetic Information Processing;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Enzyme Families;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Membrane Transport;Metabolism': 1,\n",
       " 'Enzyme Families;Genetic Information Processing;Nucleotide Metabolism': 1,\n",
       " 'Genetic Information Processing;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Energy Metabolism;Genetic Information Processing': 1}"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{k: v for k, v in sorted(kegg.items(), key=lambda x: x[1], reverse=True)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'unknown': 512851,\n",
       " 'Enzyme Families': 22267,\n",
       " 'Membrane Transport': 72775,\n",
       " 'Infectious Diseases': 1603,\n",
       " 'Amino Acid Metabolism': 56274,\n",
       " 'Carbohydrate Metabolism': 90254,\n",
       " 'Cellular Processes and Signaling': 61821,\n",
       " 'Metabolism': 54868,\n",
       " 'Signaling Molecules and Interaction': 2097,\n",
       " 'Cellular Processes and Signaling;Signaling Molecules and Interaction': 1,\n",
       " 'Replication and Repair': 46526,\n",
       " 'Lipid Metabolism': 13502,\n",
       " 'Genetic Information Processing': 49299,\n",
       " 'Metabolism of Terpenoids and Polyketides': 8769,\n",
       " 'Glycan Biosynthesis and Metabolism': 13240,\n",
       " 'Amino Acid Metabolism;Genetic Information Processing': 14,\n",
       " 'Membrane Transport;Signaling Molecules and Interaction': 1,\n",
       " 'Metabolism of Other Amino Acids': 4782,\n",
       " 'Poorly Characterized': 61400,\n",
       " 'Signal Transduction': 16494,\n",
       " 'Transcription': 36173,\n",
       " 'Genetic Information Processing;Replication and Repair': 19,\n",
       " 'Metabolism of Cofactors and Vitamins': 26134,\n",
       " 'Carbohydrate Metabolism;Metabolism': 278,\n",
       " 'Translation': 46951,\n",
       " 'Nucleotide Metabolism': 34335,\n",
       " 'Energy Metabolism': 27446,\n",
       " 'Folding, Sorting and Degradation': 17248,\n",
       " 'Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism': 32,\n",
       " 'Xenobiotics Biodegradation and Metabolism': 3118,\n",
       " 'Carbohydrate Metabolism;Enzyme Families': 2,\n",
       " 'Cell Growth and Death': 524,\n",
       " 'Metabolism;Signaling Molecules and Interaction': 1,\n",
       " 'Enzyme Families;Metabolism': 119,\n",
       " 'Cellular Processes and Signaling;Metabolism': 60,\n",
       " 'Cellular Processes and Signaling;Poorly Characterized': 23,\n",
       " 'Enzyme Families;Signal Transduction': 2,\n",
       " 'Nucleotide Metabolism;Replication and Repair': 6,\n",
       " 'Biosynthesis of Other Secondary Metabolites': 1236,\n",
       " 'Metabolism;Signal Transduction': 33,\n",
       " 'Cell Motility': 1655,\n",
       " 'Cellular Processes and Signaling;Enzyme Families': 14,\n",
       " 'Metabolism;Replication and Repair': 4,\n",
       " 'Carbohydrate Metabolism;Poorly Characterized': 21,\n",
       " 'Signal Transduction;Transcription': 57,\n",
       " 'Poorly Characterized;Translation': 2,\n",
       " 'Genetic Information Processing;Transcription': 2,\n",
       " 'Cellular Processes and Signaling;Lipid Metabolism': 3,\n",
       " 'Folding, Sorting and Degradation;Infectious Diseases': 1,\n",
       " 'Infectious Diseases;Poorly Characterized': 1,\n",
       " 'Carbohydrate Metabolism;Energy Metabolism': 18,\n",
       " 'Amino Acid Metabolism;Infectious Diseases': 1,\n",
       " 'Glycan Biosynthesis and Metabolism;Metabolism': 50,\n",
       " 'Amino Acid Metabolism;Cellular Processes and Signaling': 26,\n",
       " 'Cellular Processes and Signaling;Transcription': 2,\n",
       " 'Transport and Catabolism': 279,\n",
       " 'Cell Growth and Death;Signal Transduction': 18,\n",
       " 'Cellular Processes and Signaling;Metabolism of Other Amino Acids': 2,\n",
       " 'Cellular Processes and Signaling;Enzyme Families;Metabolism': 1,\n",
       " 'Excretory System': 56,\n",
       " 'Metabolism;Nucleotide Metabolism': 9,\n",
       " 'Metabolism;Metabolism of Other Amino Acids': 7,\n",
       " 'Genetic Information Processing;Metabolism': 12,\n",
       " 'Infectious Diseases;Metabolism': 1,\n",
       " 'Folding, Sorting and Degradation;Translation': 11,\n",
       " 'Enzyme Families;Metabolism;Signal Transduction': 1,\n",
       " 'Endocrine System': 149,\n",
       " 'Amino Acid Metabolism;Poorly Characterized': 16,\n",
       " 'Lipid Metabolism;Poorly Characterized': 7,\n",
       " 'Cellular Processes and Signaling;Membrane Transport': 3,\n",
       " 'Carbohydrate Metabolism;Lipid Metabolism': 18,\n",
       " 'Enzyme Families;Folding, Sorting and Degradation': 3,\n",
       " 'Glycan Biosynthesis and Metabolism;Metabolism of Terpenoids and Polyketides': 1,\n",
       " 'Carbohydrate Metabolism;Metabolism of Terpenoids and Polyketides': 14,\n",
       " 'Cellular Processes and Signaling;Signal Transduction': 15,\n",
       " 'Cell Growth and Death;Metabolism': 4,\n",
       " 'Membrane Transport;Replication and Repair': 2,\n",
       " 'Membrane Transport;Poorly Characterized': 1,\n",
       " 'Energy Metabolism;Folding, Sorting and Degradation;Metabolism': 1,\n",
       " 'Energy Metabolism;Metabolism': 10,\n",
       " 'Cellular Processes and Signaling;Infectious Diseases': 1,\n",
       " 'Genetic Information Processing;Membrane Transport': 12,\n",
       " 'Enzyme Families;Metabolism;unknown': 2,\n",
       " 'Cellular Processes and Signaling;Glycan Biosynthesis and Metabolism': 18,\n",
       " 'Amino Acid Metabolism;Metabolism': 37,\n",
       " 'Metabolism of Cofactors and Vitamins;Poorly Characterized': 8,\n",
       " 'Metabolism;Metabolism of Cofactors and Vitamins': 13,\n",
       " 'Amino Acid Metabolism;Enzyme Families': 5,\n",
       " 'Glycan Biosynthesis and Metabolism;Lipid Metabolism': 1,\n",
       " 'Replication and Repair;Transcription': 18,\n",
       " 'Immune System Diseases': 2,\n",
       " 'Lipid Metabolism;Metabolism of Terpenoids and Polyketides': 1,\n",
       " 'Carbohydrate Metabolism;Cellular Processes and Signaling': 12,\n",
       " 'Enzyme Families;Transport and Catabolism': 2,\n",
       " 'Enzyme Families;unknown': 1,\n",
       " 'Cell Growth and Death;Enzyme Families': 1,\n",
       " 'Energy Metabolism;Xenobiotics Biodegradation and Metabolism': 5,\n",
       " 'Cellular Processes and Signaling;Nucleotide Metabolism': 1,\n",
       " 'Amino Acid Metabolism;Energy Metabolism': 8,\n",
       " 'Carbohydrate Metabolism;Membrane Transport': 9,\n",
       " 'Biosynthesis of Other Secondary Metabolites;Lipid Metabolism': 2,\n",
       " 'Cellular Processes and Signaling;Replication and Repair': 5,\n",
       " 'Metabolism of Other Amino Acids;Xenobiotics Biodegradation and Metabolism': 1,\n",
       " 'Carbohydrate Metabolism;Metabolism of Cofactors and Vitamins': 12,\n",
       " 'Cellular Processes and Signaling;unknown': 1,\n",
       " 'Nucleotide Metabolism;Poorly Characterized': 7,\n",
       " 'Metabolism;unknown': 2,\n",
       " 'Carbohydrate Metabolism;Nucleotide Metabolism': 7,\n",
       " 'Metabolism of Other Amino Acids;Transcription': 3,\n",
       " 'Replication and Repair;Signal Transduction': 7,\n",
       " 'Energy Metabolism;Nucleotide Metabolism': 8,\n",
       " 'Cellular Processes and Signaling;Folding, Sorting and Degradation': 6,\n",
       " 'Carbohydrate Metabolism;Transcription': 17,\n",
       " 'Amino Acid Metabolism;Carbohydrate Metabolism': 9,\n",
       " 'Enzyme Families;Replication and Repair': 4,\n",
       " 'Carbohydrate Metabolism;Signaling Molecules and Interaction': 1,\n",
       " 'Lipid Metabolism;Replication and Repair': 1,\n",
       " 'Enzyme Families;Signaling Molecules and Interaction': 1,\n",
       " 'Folding, Sorting and Degradation;Membrane Transport': 1,\n",
       " 'Lipid Metabolism;Metabolism': 24,\n",
       " 'Metabolism;Metabolism of Terpenoids and Polyketides': 3,\n",
       " 'Folding, Sorting and Degradation;Genetic Information Processing': 14,\n",
       " 'Carbohydrate Metabolism;Replication and Repair': 2,\n",
       " 'Metabolism of Cofactors and Vitamins;Xenobiotics Biodegradation and Metabolism': 2,\n",
       " 'Folding, Sorting and Degradation;Genetic Information Processing;Translation': 1,\n",
       " 'Metabolism;Poorly Characterized': 5,\n",
       " 'Carbohydrate Metabolism;Glycan Biosynthesis and Metabolism;Metabolism': 1,\n",
       " 'Carbohydrate Metabolism;Transport and Catabolism': 1,\n",
       " 'Cell Motility;Membrane Transport': 1,\n",
       " 'Folding, Sorting and Degradation;Metabolism': 2,\n",
       " 'Lipid Metabolism;Xenobiotics Biodegradation and Metabolism': 9,\n",
       " 'Metabolism;Translation': 4,\n",
       " 'Metabolism;Transcription': 1,\n",
       " 'Lipid Metabolism;Transcription': 1,\n",
       " 'Poorly Characterized;Signal Transduction': 1,\n",
       " 'Carbohydrate Metabolism;Cellular Processes and Signaling;Metabolism': 1,\n",
       " 'Metabolism;Xenobiotics Biodegradation and Metabolism': 7,\n",
       " 'Carbohydrate Metabolism;Signal Transduction': 3,\n",
       " 'Cell Motility;Poorly Characterized': 1,\n",
       " 'Amino Acid Metabolism;Xenobiotics Biodegradation and Metabolism': 2,\n",
       " 'Folding, Sorting and Degradation;Metabolism of Cofactors and Vitamins': 5,\n",
       " 'Amino Acid Metabolism;Transcription': 1,\n",
       " 'Membrane Transport;Signal Transduction': 4,\n",
       " 'Membrane Transport;Transcription': 1,\n",
       " 'Amino Acid Metabolism;Metabolism of Cofactors and Vitamins': 5,\n",
       " 'Metabolism of Terpenoids and Polyketides;Translation': 2,\n",
       " 'Poorly Characterized;Xenobiotics Biodegradation and Metabolism': 3,\n",
       " 'Metabolism of Cofactors and Vitamins;Metabolism of Other Amino Acids': 1,\n",
       " 'Biosynthesis of Other Secondary Metabolites;Xenobiotics Biodegradation and Metabolism': 2,\n",
       " 'Carbohydrate Metabolism;Metabolism;Metabolism of Terpenoids and Polyketides': 1,\n",
       " 'Folding, Sorting and Degradation;Poorly Characterized': 1,\n",
       " 'Energy Metabolism;Metabolism of Other Amino Acids': 1,\n",
       " 'Metabolism of Other Amino Acids;Nucleotide Metabolism': 1,\n",
       " 'Genetic Information Processing;Xenobiotics Biodegradation and Metabolism': 1,\n",
       " 'Cellular Processes and Signaling;Metabolism of Terpenoids and Polyketides': 1,\n",
       " 'Environmental Adaptation': 2,\n",
       " 'Enzyme Families;Glycan Biosynthesis and Metabolism': 2,\n",
       " 'Energy Metabolism;Signal Transduction': 1,\n",
       " 'Cellular Processes and Signaling;Metabolism of Cofactors and Vitamins': 4,\n",
       " 'Cellular Processes and Signaling;Genetic Information Processing': 3,\n",
       " 'Signal Transduction;unknown': 2,\n",
       " 'Amino Acid Metabolism;Glycan Biosynthesis and Metabolism': 1,\n",
       " 'Cellular Processes and Signaling;Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Membrane Transport;Nucleotide Metabolism': 2,\n",
       " 'Metabolism of Cofactors and Vitamins;unknown': 1,\n",
       " 'Amino Acid Metabolism;Lipid Metabolism': 1,\n",
       " 'Enzyme Families;Nucleotide Metabolism': 12,\n",
       " 'Metabolism of Cofactors and Vitamins;Transcription': 3,\n",
       " 'Lipid Metabolism;Metabolism of Cofactors and Vitamins': 1,\n",
       " 'Genetic Information Processing;Translation': 1,\n",
       " 'Enzyme Families;Transcription': 1,\n",
       " 'Biosynthesis of Other Secondary Metabolites;Metabolism of Cofactors and Vitamins': 1,\n",
       " 'Metabolism of Cofactors and Vitamins;Metabolism of Terpenoids and Polyketides': 2,\n",
       " 'Amino Acid Metabolism;Metabolism of Terpenoids and Polyketides': 1,\n",
       " 'Digestive System': 49,\n",
       " 'Nucleotide Metabolism;Xenobiotics Biodegradation and Metabolism': 4,\n",
       " 'Cellular Processes and Signaling;Energy Metabolism;Metabolism': 1,\n",
       " 'Genetic Information Processing;Nucleotide Metabolism': 13,\n",
       " 'Cancers': 52,\n",
       " 'Enzyme Families;Membrane Transport': 1,\n",
       " 'Amino Acid Metabolism;Translation': 2,\n",
       " 'Carbohydrate Metabolism;Folding, Sorting and Degradation;Poorly Characterized': 1,\n",
       " 'Enzyme Families;Genetic Information Processing': 4,\n",
       " 'Cellular Processes and Signaling;Energy Metabolism': 1,\n",
       " 'Digestive System;Metabolism': 4,\n",
       " 'Carbohydrate Metabolism;Folding, Sorting and Degradation': 1,\n",
       " 'Glycan Biosynthesis and Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Genetic Information Processing;Glycan Biosynthesis and Metabolism': 6,\n",
       " 'Genetic Information Processing;Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Energy Metabolism;Replication and Repair': 2,\n",
       " 'Energy Metabolism;Poorly Characterized': 3,\n",
       " 'Transcription;Translation': 1,\n",
       " 'Enzyme Families;Genetic Information Processing;Glycan Biosynthesis and Metabolism': 3,\n",
       " 'Enzyme Families;Genetic Information Processing;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Enzyme Families;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Membrane Transport;Metabolism': 1,\n",
       " 'Enzyme Families;Genetic Information Processing;Nucleotide Metabolism': 1,\n",
       " 'Genetic Information Processing;Glycan Biosynthesis and Metabolism;Nucleotide Metabolism': 1,\n",
       " 'Energy Metabolism;Genetic Information Processing': 1}"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pd.Series(mydictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "numbers = [1, 2, 2, 2, 4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 4]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set(numbers))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Medidas de tendencia central"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.4.0'"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1231795176"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Length'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "102"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Length'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24615"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Length'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "958.116782121306"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Length'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "819.0"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Length'].median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    219\n",
       "Name: Gene Length, dtype: int64"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Length'].mode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "728.536548632577"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Length'].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "530765.5026934672"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Length'].var()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.25     453.0\n",
       "0.50     819.0\n",
       "0.75    1257.0\n",
       "Name: Gene Length, dtype: float64"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Length'].quantile([.25, .5, .75])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Funciones utilizadas el día de hoy**\n",
    "\n",
    "* `df.shape`\n",
    "* `df.dtypes`\n",
    "* `df.info()`\n",
    "* `df.head()`\n",
    "* `df.tail()`\n",
    "* `df.describe()`\n",
    "* `df.T`\n",
    "* `df.index`\n",
    "* `df.columns`\n",
    "* `df.set_index()`\n",
    "* `df.sort_index()`\n",
    "* `df.sort_values()`\n",
    "* `df.to_csv()`\n",
    "* `df.sum()`\n",
    "* `df.min()`\n",
    "* `df.max()`\n",
    "* `df.mean()`\n",
    "* `df.median()`\n",
    "* `df.mode()`\n",
    "* `df.std()`\n",
    "* `df.var()`\n",
    "* `df.quantile()`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtrado de columnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene Name', 'Gene Length', 'Gene Completeness', 'Cohort Origin',\n",
       "       'Taxonomic Annotation(Phylum Level)',\n",
       "       'Taxonomic Annotation(Genus Level)', 'KEGG Annotation',\n",
       "       'eggNOG Annotation', 'Sample Occurence Frequency',\n",
       "       'Individual Occurence Frequency', 'KEGG Functional Categories',\n",
       "       'eggNOG Functional Categories', 'Cohort Assembled'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     SP\n",
       "1     SP\n",
       "2    CHN\n",
       "3     SP\n",
       "4    EUR\n",
       "Name: Cohort Origin, dtype: object"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Cohort Origin'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          False\n",
       "1          False\n",
       "2          False\n",
       "3          False\n",
       "4           True\n",
       "           ...  \n",
       "1285637    False\n",
       "1285638    False\n",
       "1285639    False\n",
       "1285640    False\n",
       "1285641    False\n",
       "Name: Cohort Origin, Length: 1285642, dtype: bool"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Cohort Origin'] == 'EUR'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MH0373_GL0094743</td>\n",
       "      <td>20511</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Coprobacillus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.332281</td>\n",
       "      <td>0.338318</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MH0315_GL0002178</td>\n",
       "      <td>18975</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.043410</td>\n",
       "      <td>0.047664</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>V1.FI08_GL0038488</td>\n",
       "      <td>18753</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>bactNOG98109</td>\n",
       "      <td>0.407261</td>\n",
       "      <td>0.388785</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>V1.CD20-0_GL0037007</td>\n",
       "      <td>18738</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.421468</td>\n",
       "      <td>0.402804</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285323</th>\n",
       "      <td>V1.UC55-4_GL0094986</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Ruminococcus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285324</th>\n",
       "      <td>V1.UC55-4_GL0242705</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Roseburia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.465667</td>\n",
       "      <td>0.467290</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285325</th>\n",
       "      <td>V1.UC58-0_GL0103882</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Subdoligranulum</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.037096</td>\n",
       "      <td>0.038318</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285326</th>\n",
       "      <td>V1.UC60-0_GL0143314</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.006314</td>\n",
       "      <td>0.006542</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285327</th>\n",
       "      <td>V1.UC62-0_GL0031764</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG283857</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>797748 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "4           MH0427_GL0087973        20775          Complete           EUR   \n",
       "6           MH0373_GL0094743        20511          Complete           EUR   \n",
       "8           MH0315_GL0002178        18975          Complete           EUR   \n",
       "12         V1.FI08_GL0038488        18753          Complete           EUR   \n",
       "13       V1.CD20-0_GL0037007        18738          Complete           EUR   \n",
       "...                      ...          ...               ...           ...   \n",
       "1285323  V1.UC55-4_GL0094986          102          Complete           EUR   \n",
       "1285324  V1.UC55-4_GL0242705          102          Complete           EUR   \n",
       "1285325  V1.UC58-0_GL0103882          102          Complete           EUR   \n",
       "1285326  V1.UC60-0_GL0143314          102          Complete           EUR   \n",
       "1285327  V1.UC62-0_GL0031764          102          Complete           EUR   \n",
       "\n",
       "        Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "4                           Proteobacteria                       Escherichia   \n",
       "6                               Firmicutes                     Coprobacillus   \n",
       "8                           Proteobacteria                       Escherichia   \n",
       "12                          Proteobacteria                           unknown   \n",
       "13                          Proteobacteria                           unknown   \n",
       "...                                    ...                               ...   \n",
       "1285323                         Firmicutes                      Ruminococcus   \n",
       "1285324                         Firmicutes                         Roseburia   \n",
       "1285325                         Firmicutes                   Subdoligranulum   \n",
       "1285326                         Firmicutes                  Faecalibacterium   \n",
       "1285327                         Firmicutes                           unknown   \n",
       "\n",
       "        KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "4                K01317          NOG12793                    0.098658   \n",
       "6               unknown           unknown                    0.332281   \n",
       "8                K01317          NOG12793                    0.043410   \n",
       "12              unknown      bactNOG98109                    0.407261   \n",
       "13              unknown          NOG12793                    0.421468   \n",
       "...                 ...               ...                         ...   \n",
       "1285323         unknown           unknown                    0.000789   \n",
       "1285324         unknown           unknown                    0.465667   \n",
       "1285325         unknown           unknown                    0.037096   \n",
       "1285326         unknown           unknown                    0.006314   \n",
       "1285327         unknown         NOG283857                    0.000789   \n",
       "\n",
       "         Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "4                              0.104673            Enzyme Families   \n",
       "6                              0.338318                    unknown   \n",
       "8                              0.047664            Enzyme Families   \n",
       "12                             0.388785                    unknown   \n",
       "13                             0.402804                    unknown   \n",
       "...                                 ...                        ...   \n",
       "1285323                        0.000000                    unknown   \n",
       "1285324                        0.467290                    unknown   \n",
       "1285325                        0.038318                    unknown   \n",
       "1285326                        0.006542                    unknown   \n",
       "1285327                        0.000935                    unknown   \n",
       "\n",
       "        eggNOG Functional Categories Cohort Assembled  \n",
       "4                   Function unknown      EUR;CHN;USA  \n",
       "6                            unknown      EUR;CHN;USA  \n",
       "8                   Function unknown      EUR;CHN;USA  \n",
       "12                  Function unknown      EUR;CHN;USA  \n",
       "13                  Function unknown      EUR;CHN;USA  \n",
       "...                              ...              ...  \n",
       "1285323                      unknown              EUR  \n",
       "1285324                      unknown              EUR  \n",
       "1285325                      unknown              EUR  \n",
       "1285326                      unknown              EUR  \n",
       "1285327                      unknown              EUR  \n",
       "\n",
       "[797748 rows x 13 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Cohort Origin'] == 'EUR']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MH0373_GL0094743</td>\n",
       "      <td>20511</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Coprobacillus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.332281</td>\n",
       "      <td>0.338318</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MH0315_GL0002178</td>\n",
       "      <td>18975</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.043410</td>\n",
       "      <td>0.047664</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>V1.FI08_GL0038488</td>\n",
       "      <td>18753</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>bactNOG98109</td>\n",
       "      <td>0.407261</td>\n",
       "      <td>0.388785</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>V1.CD20-0_GL0037007</td>\n",
       "      <td>18738</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.421468</td>\n",
       "      <td>0.402804</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797743</th>\n",
       "      <td>V1.UC55-4_GL0094986</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Ruminococcus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797744</th>\n",
       "      <td>V1.UC55-4_GL0242705</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Roseburia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.465667</td>\n",
       "      <td>0.467290</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797745</th>\n",
       "      <td>V1.UC58-0_GL0103882</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Subdoligranulum</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.037096</td>\n",
       "      <td>0.038318</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797746</th>\n",
       "      <td>V1.UC60-0_GL0143314</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.006314</td>\n",
       "      <td>0.006542</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797747</th>\n",
       "      <td>V1.UC62-0_GL0031764</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG283857</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>797748 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "0          MH0427_GL0087973        20775          Complete           EUR   \n",
       "1          MH0373_GL0094743        20511          Complete           EUR   \n",
       "2          MH0315_GL0002178        18975          Complete           EUR   \n",
       "3         V1.FI08_GL0038488        18753          Complete           EUR   \n",
       "4       V1.CD20-0_GL0037007        18738          Complete           EUR   \n",
       "...                     ...          ...               ...           ...   \n",
       "797743  V1.UC55-4_GL0094986          102          Complete           EUR   \n",
       "797744  V1.UC55-4_GL0242705          102          Complete           EUR   \n",
       "797745  V1.UC58-0_GL0103882          102          Complete           EUR   \n",
       "797746  V1.UC60-0_GL0143314          102          Complete           EUR   \n",
       "797747  V1.UC62-0_GL0031764          102          Complete           EUR   \n",
       "\n",
       "       Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "0                          Proteobacteria                       Escherichia   \n",
       "1                              Firmicutes                     Coprobacillus   \n",
       "2                          Proteobacteria                       Escherichia   \n",
       "3                          Proteobacteria                           unknown   \n",
       "4                          Proteobacteria                           unknown   \n",
       "...                                   ...                               ...   \n",
       "797743                         Firmicutes                      Ruminococcus   \n",
       "797744                         Firmicutes                         Roseburia   \n",
       "797745                         Firmicutes                   Subdoligranulum   \n",
       "797746                         Firmicutes                  Faecalibacterium   \n",
       "797747                         Firmicutes                           unknown   \n",
       "\n",
       "       KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "0               K01317          NOG12793                    0.098658   \n",
       "1              unknown           unknown                    0.332281   \n",
       "2               K01317          NOG12793                    0.043410   \n",
       "3              unknown      bactNOG98109                    0.407261   \n",
       "4              unknown          NOG12793                    0.421468   \n",
       "...                ...               ...                         ...   \n",
       "797743         unknown           unknown                    0.000789   \n",
       "797744         unknown           unknown                    0.465667   \n",
       "797745         unknown           unknown                    0.037096   \n",
       "797746         unknown           unknown                    0.006314   \n",
       "797747         unknown         NOG283857                    0.000789   \n",
       "\n",
       "        Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "0                             0.104673            Enzyme Families   \n",
       "1                             0.338318                    unknown   \n",
       "2                             0.047664            Enzyme Families   \n",
       "3                             0.388785                    unknown   \n",
       "4                             0.402804                    unknown   \n",
       "...                                ...                        ...   \n",
       "797743                        0.000000                    unknown   \n",
       "797744                        0.467290                    unknown   \n",
       "797745                        0.038318                    unknown   \n",
       "797746                        0.006542                    unknown   \n",
       "797747                        0.000935                    unknown   \n",
       "\n",
       "       eggNOG Functional Categories Cohort Assembled  \n",
       "0                  Function unknown      EUR;CHN;USA  \n",
       "1                           unknown      EUR;CHN;USA  \n",
       "2                  Function unknown      EUR;CHN;USA  \n",
       "3                  Function unknown      EUR;CHN;USA  \n",
       "4                  Function unknown      EUR;CHN;USA  \n",
       "...                             ...              ...  \n",
       "797743                      unknown              EUR  \n",
       "797744                      unknown              EUR  \n",
       "797745                      unknown              EUR  \n",
       "797746                      unknown              EUR  \n",
       "797747                      unknown              EUR  \n",
       "\n",
       "[797748 rows x 13 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Cohort Origin'] == 'EUR'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>24615</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Weissella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.066298</td>\n",
       "      <td>0.072897</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.095501</td>\n",
       "      <td>0.096262</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285637</th>\n",
       "      <td>665954.HMPREF1017_02554</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.089187</td>\n",
       "      <td>0.077570</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285638</th>\n",
       "      <td>753642.ECNC101_09444</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285639</th>\n",
       "      <td>768724.HMPREF9131_0965</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Peptoniphilus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285640</th>\n",
       "      <td>866771.HMPREF9296_1721</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285641</th>\n",
       "      <td>926026.ECO5101_18612</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>proNOG05789</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1285642 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Gene Name  Gene Length Gene Completeness  \\\n",
       "0        911104.WcibK1_010100007220        24615          Complete   \n",
       "1                  585054.EFER_0542        21669          Complete   \n",
       "2                SZEY-48A_GL0052647        20778          Complete   \n",
       "3           1048689.ECO55CA74_02930        20778          Complete   \n",
       "4                  MH0427_GL0087973        20775          Complete   \n",
       "...                             ...          ...               ...   \n",
       "1285637     665954.HMPREF1017_02554          102          Complete   \n",
       "1285638        753642.ECNC101_09444          102          Complete   \n",
       "1285639      768724.HMPREF9131_0965          102          Complete   \n",
       "1285640      866771.HMPREF9296_1721          102          Complete   \n",
       "1285641        926026.ECO5101_18612          102          Complete   \n",
       "\n",
       "        Cohort Origin Taxonomic Annotation(Phylum Level)  \\\n",
       "0                  SP                         Firmicutes   \n",
       "1                  SP                     Proteobacteria   \n",
       "2                 CHN                     Proteobacteria   \n",
       "3                  SP                     Proteobacteria   \n",
       "4                 EUR                     Proteobacteria   \n",
       "...               ...                                ...   \n",
       "1285637            SP                      Bacteroidetes   \n",
       "1285638            SP                     Proteobacteria   \n",
       "1285639            SP                         Firmicutes   \n",
       "1285640            SP                      Bacteroidetes   \n",
       "1285641            SP                     Proteobacteria   \n",
       "\n",
       "        Taxonomic Annotation(Genus Level) KEGG Annotation eggNOG Annotation  \\\n",
       "0                               Weissella         unknown          NOG12793   \n",
       "1                             Escherichia         unknown          NOG12793   \n",
       "2                             Escherichia          K01317          NOG12793   \n",
       "3                             Escherichia          K01317          NOG12793   \n",
       "4                             Escherichia          K01317          NOG12793   \n",
       "...                                   ...             ...               ...   \n",
       "1285637                       Bacteroides         unknown           unknown   \n",
       "1285638                       Escherichia         unknown           unknown   \n",
       "1285639                     Peptoniphilus         unknown           unknown   \n",
       "1285640                        Prevotella         unknown           unknown   \n",
       "1285641                       Escherichia         unknown       proNOG05789   \n",
       "\n",
       "         Sample Occurence Frequency  Individual Occurence Frequency  \\\n",
       "0                          0.066298                        0.072897   \n",
       "1                          0.058406                        0.062617   \n",
       "2                          0.203631                        0.212150   \n",
       "3                          0.095501                        0.096262   \n",
       "4                          0.098658                        0.104673   \n",
       "...                             ...                             ...   \n",
       "1285637                    0.089187                        0.077570   \n",
       "1285638                    0.001579                        0.000935   \n",
       "1285639                    0.000789                        0.000935   \n",
       "1285640                    0.000789                        0.000935   \n",
       "1285641                    0.001579                        0.001869   \n",
       "\n",
       "        KEGG Functional Categories eggNOG Functional Categories  \\\n",
       "0                          unknown             Function unknown   \n",
       "1                          unknown             Function unknown   \n",
       "2                  Enzyme Families             Function unknown   \n",
       "3                  Enzyme Families             Function unknown   \n",
       "4                  Enzyme Families             Function unknown   \n",
       "...                            ...                          ...   \n",
       "1285637                    unknown                      unknown   \n",
       "1285638                    unknown                      unknown   \n",
       "1285639                    unknown                      unknown   \n",
       "1285640                    unknown                      unknown   \n",
       "1285641                    unknown             Function unknown   \n",
       "\n",
       "        Cohort Assembled  \n",
       "0                    EUR  \n",
       "1                EUR;CHN  \n",
       "2                EUR;CHN  \n",
       "3                EUR;CHN  \n",
       "4            EUR;CHN;USA  \n",
       "...                  ...  \n",
       "1285637              NaN  \n",
       "1285638              NaN  \n",
       "1285639              NaN  \n",
       "1285640              NaN  \n",
       "1285641              NaN  \n",
       "\n",
       "[1285642 rows x 13 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Gene Completeness'] == 'Complete']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene Name', 'Gene Length', 'Gene Completeness', 'Cohort Origin',\n",
       "       'Taxonomic Annotation(Phylum Level)',\n",
       "       'Taxonomic Annotation(Genus Level)', 'KEGG Annotation',\n",
       "       'eggNOG Annotation', 'Sample Occurence Frequency',\n",
       "       'Individual Occurence Frequency', 'KEGG Functional Categories',\n",
       "       'eggNOG Functional Categories', 'Cohort Assembled'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MH0373_GL0094743</td>\n",
       "      <td>20511</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Coprobacillus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.332281</td>\n",
       "      <td>0.338318</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MH0315_GL0002178</td>\n",
       "      <td>18975</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.043410</td>\n",
       "      <td>0.047664</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>V1.FI08_GL0038488</td>\n",
       "      <td>18753</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>bactNOG98109</td>\n",
       "      <td>0.407261</td>\n",
       "      <td>0.388785</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>V1.CD20-0_GL0037007</td>\n",
       "      <td>18738</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.421468</td>\n",
       "      <td>0.402804</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285323</th>\n",
       "      <td>V1.UC55-4_GL0094986</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Ruminococcus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285324</th>\n",
       "      <td>V1.UC55-4_GL0242705</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Roseburia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.465667</td>\n",
       "      <td>0.467290</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285325</th>\n",
       "      <td>V1.UC58-0_GL0103882</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Subdoligranulum</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.037096</td>\n",
       "      <td>0.038318</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285326</th>\n",
       "      <td>V1.UC60-0_GL0143314</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.006314</td>\n",
       "      <td>0.006542</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285327</th>\n",
       "      <td>V1.UC62-0_GL0031764</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG283857</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>797748 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "4           MH0427_GL0087973        20775          Complete           EUR   \n",
       "6           MH0373_GL0094743        20511          Complete           EUR   \n",
       "8           MH0315_GL0002178        18975          Complete           EUR   \n",
       "12         V1.FI08_GL0038488        18753          Complete           EUR   \n",
       "13       V1.CD20-0_GL0037007        18738          Complete           EUR   \n",
       "...                      ...          ...               ...           ...   \n",
       "1285323  V1.UC55-4_GL0094986          102          Complete           EUR   \n",
       "1285324  V1.UC55-4_GL0242705          102          Complete           EUR   \n",
       "1285325  V1.UC58-0_GL0103882          102          Complete           EUR   \n",
       "1285326  V1.UC60-0_GL0143314          102          Complete           EUR   \n",
       "1285327  V1.UC62-0_GL0031764          102          Complete           EUR   \n",
       "\n",
       "        Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "4                           Proteobacteria                       Escherichia   \n",
       "6                               Firmicutes                     Coprobacillus   \n",
       "8                           Proteobacteria                       Escherichia   \n",
       "12                          Proteobacteria                           unknown   \n",
       "13                          Proteobacteria                           unknown   \n",
       "...                                    ...                               ...   \n",
       "1285323                         Firmicutes                      Ruminococcus   \n",
       "1285324                         Firmicutes                         Roseburia   \n",
       "1285325                         Firmicutes                   Subdoligranulum   \n",
       "1285326                         Firmicutes                  Faecalibacterium   \n",
       "1285327                         Firmicutes                           unknown   \n",
       "\n",
       "        KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "4                K01317          NOG12793                    0.098658   \n",
       "6               unknown           unknown                    0.332281   \n",
       "8                K01317          NOG12793                    0.043410   \n",
       "12              unknown      bactNOG98109                    0.407261   \n",
       "13              unknown          NOG12793                    0.421468   \n",
       "...                 ...               ...                         ...   \n",
       "1285323         unknown           unknown                    0.000789   \n",
       "1285324         unknown           unknown                    0.465667   \n",
       "1285325         unknown           unknown                    0.037096   \n",
       "1285326         unknown           unknown                    0.006314   \n",
       "1285327         unknown         NOG283857                    0.000789   \n",
       "\n",
       "         Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "4                              0.104673            Enzyme Families   \n",
       "6                              0.338318                    unknown   \n",
       "8                              0.047664            Enzyme Families   \n",
       "12                             0.388785                    unknown   \n",
       "13                             0.402804                    unknown   \n",
       "...                                 ...                        ...   \n",
       "1285323                        0.000000                    unknown   \n",
       "1285324                        0.467290                    unknown   \n",
       "1285325                        0.038318                    unknown   \n",
       "1285326                        0.006542                    unknown   \n",
       "1285327                        0.000935                    unknown   \n",
       "\n",
       "        eggNOG Functional Categories Cohort Assembled  \n",
       "4                   Function unknown      EUR;CHN;USA  \n",
       "6                            unknown      EUR;CHN;USA  \n",
       "8                   Function unknown      EUR;CHN;USA  \n",
       "12                  Function unknown      EUR;CHN;USA  \n",
       "13                  Function unknown      EUR;CHN;USA  \n",
       "...                              ...              ...  \n",
       "1285323                      unknown              EUR  \n",
       "1285324                      unknown              EUR  \n",
       "1285325                      unknown              EUR  \n",
       "1285326                      unknown              EUR  \n",
       "1285327                      unknown              EUR  \n",
       "\n",
       "[797748 rows x 13 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.query('`Gene Completeness` == \"Complete\" & `Cohort Origin` == \"EUR\"' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2713</th>\n",
       "      <td>MH0025_GL0055549</td>\n",
       "      <td>4959</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K14326</td>\n",
       "      <td>COG1112</td>\n",
       "      <td>0.032360</td>\n",
       "      <td>0.037383</td>\n",
       "      <td>Translation</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1020000</th>\n",
       "      <td>MH0008_GL0018223</td>\n",
       "      <td>393</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K02996</td>\n",
       "      <td>COG0103</td>\n",
       "      <td>0.628256</td>\n",
       "      <td>0.649533</td>\n",
       "      <td>Translation</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1026977</th>\n",
       "      <td>V1.CD3-0-PT_GL0051806</td>\n",
       "      <td>387</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K02994</td>\n",
       "      <td>COG0096</td>\n",
       "      <td>0.086030</td>\n",
       "      <td>0.086916</td>\n",
       "      <td>Translation</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1116958</th>\n",
       "      <td>V1.CD3-3-PN_GL0019322</td>\n",
       "      <td>294</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K07574</td>\n",
       "      <td>COG1534</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.057009</td>\n",
       "      <td>Translation</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1131123</th>\n",
       "      <td>V1.CD7-0-PN_GL0013326</td>\n",
       "      <td>279</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K02965</td>\n",
       "      <td>COG0185</td>\n",
       "      <td>0.022099</td>\n",
       "      <td>0.018692</td>\n",
       "      <td>Translation</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "2713          MH0025_GL0055549         4959          Complete           EUR   \n",
       "1020000       MH0008_GL0018223          393          Complete           EUR   \n",
       "1026977  V1.CD3-0-PT_GL0051806          387          Complete           EUR   \n",
       "1116958  V1.CD3-3-PN_GL0019322          294          Complete           EUR   \n",
       "1131123  V1.CD7-0-PN_GL0013326          279          Complete           EUR   \n",
       "\n",
       "        Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "2713                        Proteobacteria                        Salmonella   \n",
       "1020000                     Proteobacteria                        Salmonella   \n",
       "1026977                     Proteobacteria                        Salmonella   \n",
       "1116958                     Proteobacteria                        Salmonella   \n",
       "1131123                     Proteobacteria                        Salmonella   \n",
       "\n",
       "        KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "2713             K14326           COG1112                    0.032360   \n",
       "1020000          K02996           COG0103                    0.628256   \n",
       "1026977          K02994           COG0096                    0.086030   \n",
       "1116958          K07574           COG1534                    0.058406   \n",
       "1131123          K02965           COG0185                    0.022099   \n",
       "\n",
       "         Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "2713                           0.037383                Translation   \n",
       "1020000                        0.649533                Translation   \n",
       "1026977                        0.086916                Translation   \n",
       "1116958                        0.057009                Translation   \n",
       "1131123                        0.018692                Translation   \n",
       "\n",
       "                            eggNOG Functional Categories Cohort Assembled  \n",
       "2713               Replication, recombination and repair          EUR;CHN  \n",
       "1020000  Translation, ribosomal structure and biogenesis      EUR;CHN;USA  \n",
       "1026977  Translation, ribosomal structure and biogenesis              EUR  \n",
       "1116958  Translation, ribosomal structure and biogenesis              EUR  \n",
       "1131123  Translation, ribosomal structure and biogenesis              EUR  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df[\"Taxonomic Annotation(Genus Level)\"] == 'Salmonella') \n",
    "    & (df['Cohort Origin'] == 'EUR')\n",
    "    & (df['KEGG Functional Categories'] == \"Translation\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1270</th>\n",
       "      <td>MH0008_GL0021722</td>\n",
       "      <td>6342</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K06877</td>\n",
       "      <td>COG1205</td>\n",
       "      <td>0.063931</td>\n",
       "      <td>0.071028</td>\n",
       "      <td>Poorly Characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2713</th>\n",
       "      <td>MH0025_GL0055549</td>\n",
       "      <td>4959</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K14326</td>\n",
       "      <td>COG1112</td>\n",
       "      <td>0.032360</td>\n",
       "      <td>0.037383</td>\n",
       "      <td>Translation</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2807</th>\n",
       "      <td>MH0008_GL0021723</td>\n",
       "      <td>4914</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K00571</td>\n",
       "      <td>COG1002</td>\n",
       "      <td>0.035517</td>\n",
       "      <td>0.040187</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3930</th>\n",
       "      <td>T2D-77A_GL0017130</td>\n",
       "      <td>4518</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K06877</td>\n",
       "      <td>COG1205</td>\n",
       "      <td>0.030781</td>\n",
       "      <td>0.033645</td>\n",
       "      <td>Poorly Characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6668</th>\n",
       "      <td>V1.CD3-0-PT_GL0053841</td>\n",
       "      <td>4029</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K03043</td>\n",
       "      <td>COG0085</td>\n",
       "      <td>0.024467</td>\n",
       "      <td>0.018692</td>\n",
       "      <td>Nucleotide Metabolism</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276755</th>\n",
       "      <td>O2.UC24-2_GL0128465</td>\n",
       "      <td>120</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.029992</td>\n",
       "      <td>0.030841</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276963</th>\n",
       "      <td>NLM027_GL0023741</td>\n",
       "      <td>120</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG81196</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278878</th>\n",
       "      <td>MH0095_GL0092146</td>\n",
       "      <td>117</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K07497</td>\n",
       "      <td>COG2801</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.006542</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1279465</th>\n",
       "      <td>T2D-66A_GL0073432</td>\n",
       "      <td>117</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.380426</td>\n",
       "      <td>0.402804</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1284092</th>\n",
       "      <td>MH0133_GL0089869</td>\n",
       "      <td>105</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.003157</td>\n",
       "      <td>0.003738</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>935 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "1270          MH0008_GL0021722         6342          Complete           EUR   \n",
       "2713          MH0025_GL0055549         4959          Complete           EUR   \n",
       "2807          MH0008_GL0021723         4914          Complete           EUR   \n",
       "3930         T2D-77A_GL0017130         4518          Complete           CHN   \n",
       "6668     V1.CD3-0-PT_GL0053841         4029          Complete           EUR   \n",
       "...                        ...          ...               ...           ...   \n",
       "1276755    O2.UC24-2_GL0128465          120          Complete           EUR   \n",
       "1276963       NLM027_GL0023741          120          Complete           CHN   \n",
       "1278878       MH0095_GL0092146          117          Complete           EUR   \n",
       "1279465      T2D-66A_GL0073432          117          Complete           CHN   \n",
       "1284092       MH0133_GL0089869          105          Complete           EUR   \n",
       "\n",
       "        Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "1270                        Proteobacteria                        Salmonella   \n",
       "2713                        Proteobacteria                        Salmonella   \n",
       "2807                        Proteobacteria                        Salmonella   \n",
       "3930                        Proteobacteria                        Salmonella   \n",
       "6668                        Proteobacteria                        Salmonella   \n",
       "...                                    ...                               ...   \n",
       "1276755                     Proteobacteria                        Salmonella   \n",
       "1276963                     Proteobacteria                        Salmonella   \n",
       "1278878                     Proteobacteria                        Salmonella   \n",
       "1279465                     Proteobacteria                        Salmonella   \n",
       "1284092                     Proteobacteria                        Salmonella   \n",
       "\n",
       "        KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "1270             K06877           COG1205                    0.063931   \n",
       "2713             K14326           COG1112                    0.032360   \n",
       "2807             K00571           COG1002                    0.035517   \n",
       "3930             K06877           COG1205                    0.030781   \n",
       "6668             K03043           COG0085                    0.024467   \n",
       "...                 ...               ...                         ...   \n",
       "1276755         unknown           unknown                    0.029992   \n",
       "1276963         unknown          NOG81196                    0.001579   \n",
       "1278878          K07497           COG2801                    0.005525   \n",
       "1279465         unknown           unknown                    0.380426   \n",
       "1284092         unknown          NOG12793                    0.003157   \n",
       "\n",
       "         Individual Occurence Frequency      KEGG Functional Categories  \\\n",
       "1270                           0.071028            Poorly Characterized   \n",
       "2713                           0.037383                     Translation   \n",
       "2807                           0.040187  Genetic Information Processing   \n",
       "3930                           0.033645            Poorly Characterized   \n",
       "6668                           0.018692           Nucleotide Metabolism   \n",
       "...                                 ...                             ...   \n",
       "1276755                        0.030841                         unknown   \n",
       "1276963                        0.001869                         unknown   \n",
       "1278878                        0.006542  Genetic Information Processing   \n",
       "1279465                        0.402804                         unknown   \n",
       "1284092                        0.003738                         unknown   \n",
       "\n",
       "                  eggNOG Functional Categories Cohort Assembled  \n",
       "1270          General function prediction only          EUR;CHN  \n",
       "2713     Replication, recombination and repair          EUR;CHN  \n",
       "2807                        Defense mechanisms              EUR  \n",
       "3930          General function prediction only              CHN  \n",
       "6668                             Transcription          EUR;CHN  \n",
       "...                                        ...              ...  \n",
       "1276755                                unknown      EUR;CHN;USA  \n",
       "1276963                       Function unknown              CHN  \n",
       "1278878  Replication, recombination and repair              EUR  \n",
       "1279465                                unknown          CHN;USA  \n",
       "1284092                       Function unknown              EUR  \n",
       "\n",
       "[935 rows x 13 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Taxonomic Annotation(Genus Level)'].str.contains('Salmonella')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1270</th>\n",
       "      <td>MH0008_GL0021722</td>\n",
       "      <td>6342</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K06877</td>\n",
       "      <td>COG1205</td>\n",
       "      <td>0.063931</td>\n",
       "      <td>0.071028</td>\n",
       "      <td>Poorly Characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2713</th>\n",
       "      <td>MH0025_GL0055549</td>\n",
       "      <td>4959</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K14326</td>\n",
       "      <td>COG1112</td>\n",
       "      <td>0.032360</td>\n",
       "      <td>0.037383</td>\n",
       "      <td>Translation</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2807</th>\n",
       "      <td>MH0008_GL0021723</td>\n",
       "      <td>4914</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K00571</td>\n",
       "      <td>COG1002</td>\n",
       "      <td>0.035517</td>\n",
       "      <td>0.040187</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3930</th>\n",
       "      <td>T2D-77A_GL0017130</td>\n",
       "      <td>4518</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K06877</td>\n",
       "      <td>COG1205</td>\n",
       "      <td>0.030781</td>\n",
       "      <td>0.033645</td>\n",
       "      <td>Poorly Characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6668</th>\n",
       "      <td>V1.CD3-0-PT_GL0053841</td>\n",
       "      <td>4029</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K03043</td>\n",
       "      <td>COG0085</td>\n",
       "      <td>0.024467</td>\n",
       "      <td>0.018692</td>\n",
       "      <td>Nucleotide Metabolism</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276755</th>\n",
       "      <td>O2.UC24-2_GL0128465</td>\n",
       "      <td>120</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.029992</td>\n",
       "      <td>0.030841</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276963</th>\n",
       "      <td>NLM027_GL0023741</td>\n",
       "      <td>120</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG81196</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1278878</th>\n",
       "      <td>MH0095_GL0092146</td>\n",
       "      <td>117</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>K07497</td>\n",
       "      <td>COG2801</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.006542</td>\n",
       "      <td>Genetic Information Processing</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1279465</th>\n",
       "      <td>T2D-66A_GL0073432</td>\n",
       "      <td>117</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.380426</td>\n",
       "      <td>0.402804</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1284092</th>\n",
       "      <td>MH0133_GL0089869</td>\n",
       "      <td>105</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.003157</td>\n",
       "      <td>0.003738</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>935 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "1270          MH0008_GL0021722         6342          Complete           EUR   \n",
       "2713          MH0025_GL0055549         4959          Complete           EUR   \n",
       "2807          MH0008_GL0021723         4914          Complete           EUR   \n",
       "3930         T2D-77A_GL0017130         4518          Complete           CHN   \n",
       "6668     V1.CD3-0-PT_GL0053841         4029          Complete           EUR   \n",
       "...                        ...          ...               ...           ...   \n",
       "1276755    O2.UC24-2_GL0128465          120          Complete           EUR   \n",
       "1276963       NLM027_GL0023741          120          Complete           CHN   \n",
       "1278878       MH0095_GL0092146          117          Complete           EUR   \n",
       "1279465      T2D-66A_GL0073432          117          Complete           CHN   \n",
       "1284092       MH0133_GL0089869          105          Complete           EUR   \n",
       "\n",
       "        Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "1270                        Proteobacteria                        Salmonella   \n",
       "2713                        Proteobacteria                        Salmonella   \n",
       "2807                        Proteobacteria                        Salmonella   \n",
       "3930                        Proteobacteria                        Salmonella   \n",
       "6668                        Proteobacteria                        Salmonella   \n",
       "...                                    ...                               ...   \n",
       "1276755                     Proteobacteria                        Salmonella   \n",
       "1276963                     Proteobacteria                        Salmonella   \n",
       "1278878                     Proteobacteria                        Salmonella   \n",
       "1279465                     Proteobacteria                        Salmonella   \n",
       "1284092                     Proteobacteria                        Salmonella   \n",
       "\n",
       "        KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "1270             K06877           COG1205                    0.063931   \n",
       "2713             K14326           COG1112                    0.032360   \n",
       "2807             K00571           COG1002                    0.035517   \n",
       "3930             K06877           COG1205                    0.030781   \n",
       "6668             K03043           COG0085                    0.024467   \n",
       "...                 ...               ...                         ...   \n",
       "1276755         unknown           unknown                    0.029992   \n",
       "1276963         unknown          NOG81196                    0.001579   \n",
       "1278878          K07497           COG2801                    0.005525   \n",
       "1279465         unknown           unknown                    0.380426   \n",
       "1284092         unknown          NOG12793                    0.003157   \n",
       "\n",
       "         Individual Occurence Frequency      KEGG Functional Categories  \\\n",
       "1270                           0.071028            Poorly Characterized   \n",
       "2713                           0.037383                     Translation   \n",
       "2807                           0.040187  Genetic Information Processing   \n",
       "3930                           0.033645            Poorly Characterized   \n",
       "6668                           0.018692           Nucleotide Metabolism   \n",
       "...                                 ...                             ...   \n",
       "1276755                        0.030841                         unknown   \n",
       "1276963                        0.001869                         unknown   \n",
       "1278878                        0.006542  Genetic Information Processing   \n",
       "1279465                        0.402804                         unknown   \n",
       "1284092                        0.003738                         unknown   \n",
       "\n",
       "                  eggNOG Functional Categories Cohort Assembled  \n",
       "1270          General function prediction only          EUR;CHN  \n",
       "2713     Replication, recombination and repair          EUR;CHN  \n",
       "2807                        Defense mechanisms              EUR  \n",
       "3930          General function prediction only              CHN  \n",
       "6668                             Transcription          EUR;CHN  \n",
       "...                                        ...              ...  \n",
       "1276755                                unknown      EUR;CHN;USA  \n",
       "1276963                       Function unknown              CHN  \n",
       "1278878  Replication, recombination and repair              EUR  \n",
       "1279465                                unknown          CHN;USA  \n",
       "1284092                       Function unknown              EUR  \n",
       "\n",
       "[935 rows x 13 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Taxonomic Annotation(Genus Level)'].str.contains('sAlmonELLa', case=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(935, 13)"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Taxonomic Annotation(Genus Level)'].str.contains('Salmonella')].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(34019, 13)"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Taxonomic Annotation(Genus Level)'].str.contains('Escherichia')].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Salmonella|Escherichia'"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'|'.join(['Salmonella', 'Escherichia'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.095501</td>\n",
       "      <td>0.096262</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>754082.ECSTEC7V_0573</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.068666</td>\n",
       "      <td>0.073832</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285627</th>\n",
       "      <td>656444.ECNG_05213</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285628</th>\n",
       "      <td>656444.ECNG_04574</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.042620</td>\n",
       "      <td>0.049533</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285629</th>\n",
       "      <td>656444.ECNG_05118</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>COG1475</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285638</th>\n",
       "      <td>753642.ECNC101_09444</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285641</th>\n",
       "      <td>926026.ECO5101_18612</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>proNOG05789</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34954 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "1               585054.EFER_0542        21669          Complete            SP   \n",
       "2             SZEY-48A_GL0052647        20778          Complete           CHN   \n",
       "3        1048689.ECO55CA74_02930        20778          Complete            SP   \n",
       "4               MH0427_GL0087973        20775          Complete           EUR   \n",
       "5           754082.ECSTEC7V_0573        20775          Complete            SP   \n",
       "...                          ...          ...               ...           ...   \n",
       "1285627        656444.ECNG_05213          102          Complete            SP   \n",
       "1285628        656444.ECNG_04574          102          Complete            SP   \n",
       "1285629        656444.ECNG_05118          102          Complete            SP   \n",
       "1285638     753642.ECNC101_09444          102          Complete            SP   \n",
       "1285641     926026.ECO5101_18612          102          Complete            SP   \n",
       "\n",
       "        Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "1                           Proteobacteria                       Escherichia   \n",
       "2                           Proteobacteria                       Escherichia   \n",
       "3                           Proteobacteria                       Escherichia   \n",
       "4                           Proteobacteria                       Escherichia   \n",
       "5                           Proteobacteria                       Escherichia   \n",
       "...                                    ...                               ...   \n",
       "1285627                     Proteobacteria                       Escherichia   \n",
       "1285628                     Proteobacteria                       Escherichia   \n",
       "1285629                     Proteobacteria                       Escherichia   \n",
       "1285638                     Proteobacteria                       Escherichia   \n",
       "1285641                     Proteobacteria                       Escherichia   \n",
       "\n",
       "        KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "1               unknown          NOG12793                    0.058406   \n",
       "2                K01317          NOG12793                    0.203631   \n",
       "3                K01317          NOG12793                    0.095501   \n",
       "4                K01317          NOG12793                    0.098658   \n",
       "5                K01317          NOG12793                    0.068666   \n",
       "...                 ...               ...                         ...   \n",
       "1285627         unknown           unknown                    0.000789   \n",
       "1285628         unknown           unknown                    0.042620   \n",
       "1285629         unknown           COG1475                    0.000000   \n",
       "1285638         unknown           unknown                    0.001579   \n",
       "1285641         unknown       proNOG05789                    0.001579   \n",
       "\n",
       "         Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "1                              0.062617                    unknown   \n",
       "2                              0.212150            Enzyme Families   \n",
       "3                              0.096262            Enzyme Families   \n",
       "4                              0.104673            Enzyme Families   \n",
       "5                              0.073832            Enzyme Families   \n",
       "...                                 ...                        ...   \n",
       "1285627                        0.000935                    unknown   \n",
       "1285628                        0.049533                    unknown   \n",
       "1285629                        0.000000                    unknown   \n",
       "1285638                        0.000935                    unknown   \n",
       "1285641                        0.001869                    unknown   \n",
       "\n",
       "        eggNOG Functional Categories Cohort Assembled  \n",
       "1                   Function unknown          EUR;CHN  \n",
       "2                   Function unknown          EUR;CHN  \n",
       "3                   Function unknown          EUR;CHN  \n",
       "4                   Function unknown      EUR;CHN;USA  \n",
       "5                   Function unknown          EUR;CHN  \n",
       "...                              ...              ...  \n",
       "1285627                      unknown              NaN  \n",
       "1285628                      unknown              NaN  \n",
       "1285629                Transcription              NaN  \n",
       "1285638                      unknown              NaN  \n",
       "1285641             Function unknown              NaN  \n",
       "\n",
       "[34954 rows x 13 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Taxonomic Annotation(Genus Level)'].str.contains('Salmonella|Escherichia')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.095501</td>\n",
       "      <td>0.096262</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>754082.ECSTEC7V_0573</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.068666</td>\n",
       "      <td>0.073832</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285627</th>\n",
       "      <td>656444.ECNG_05213</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285628</th>\n",
       "      <td>656444.ECNG_04574</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.042620</td>\n",
       "      <td>0.049533</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285629</th>\n",
       "      <td>656444.ECNG_05118</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>COG1475</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285638</th>\n",
       "      <td>753642.ECNC101_09444</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285641</th>\n",
       "      <td>926026.ECO5101_18612</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>proNOG05789</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34954 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "1               585054.EFER_0542        21669          Complete            SP   \n",
       "2             SZEY-48A_GL0052647        20778          Complete           CHN   \n",
       "3        1048689.ECO55CA74_02930        20778          Complete            SP   \n",
       "4               MH0427_GL0087973        20775          Complete           EUR   \n",
       "5           754082.ECSTEC7V_0573        20775          Complete            SP   \n",
       "...                          ...          ...               ...           ...   \n",
       "1285627        656444.ECNG_05213          102          Complete            SP   \n",
       "1285628        656444.ECNG_04574          102          Complete            SP   \n",
       "1285629        656444.ECNG_05118          102          Complete            SP   \n",
       "1285638     753642.ECNC101_09444          102          Complete            SP   \n",
       "1285641     926026.ECO5101_18612          102          Complete            SP   \n",
       "\n",
       "        Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "1                           Proteobacteria                       Escherichia   \n",
       "2                           Proteobacteria                       Escherichia   \n",
       "3                           Proteobacteria                       Escherichia   \n",
       "4                           Proteobacteria                       Escherichia   \n",
       "5                           Proteobacteria                       Escherichia   \n",
       "...                                    ...                               ...   \n",
       "1285627                     Proteobacteria                       Escherichia   \n",
       "1285628                     Proteobacteria                       Escherichia   \n",
       "1285629                     Proteobacteria                       Escherichia   \n",
       "1285638                     Proteobacteria                       Escherichia   \n",
       "1285641                     Proteobacteria                       Escherichia   \n",
       "\n",
       "        KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "1               unknown          NOG12793                    0.058406   \n",
       "2                K01317          NOG12793                    0.203631   \n",
       "3                K01317          NOG12793                    0.095501   \n",
       "4                K01317          NOG12793                    0.098658   \n",
       "5                K01317          NOG12793                    0.068666   \n",
       "...                 ...               ...                         ...   \n",
       "1285627         unknown           unknown                    0.000789   \n",
       "1285628         unknown           unknown                    0.042620   \n",
       "1285629         unknown           COG1475                    0.000000   \n",
       "1285638         unknown           unknown                    0.001579   \n",
       "1285641         unknown       proNOG05789                    0.001579   \n",
       "\n",
       "         Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "1                              0.062617                    unknown   \n",
       "2                              0.212150            Enzyme Families   \n",
       "3                              0.096262            Enzyme Families   \n",
       "4                              0.104673            Enzyme Families   \n",
       "5                              0.073832            Enzyme Families   \n",
       "...                                 ...                        ...   \n",
       "1285627                        0.000935                    unknown   \n",
       "1285628                        0.049533                    unknown   \n",
       "1285629                        0.000000                    unknown   \n",
       "1285638                        0.000935                    unknown   \n",
       "1285641                        0.001869                    unknown   \n",
       "\n",
       "        eggNOG Functional Categories Cohort Assembled  \n",
       "1                   Function unknown          EUR;CHN  \n",
       "2                   Function unknown          EUR;CHN  \n",
       "3                   Function unknown          EUR;CHN  \n",
       "4                   Function unknown      EUR;CHN;USA  \n",
       "5                   Function unknown          EUR;CHN  \n",
       "...                              ...              ...  \n",
       "1285627                      unknown              NaN  \n",
       "1285628                      unknown              NaN  \n",
       "1285629                Transcription              NaN  \n",
       "1285638                      unknown              NaN  \n",
       "1285641             Function unknown              NaN  \n",
       "\n",
       "[34954 rows x 13 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Taxonomic Annotation(Genus Level)'].str.contains('|'.join(['Salmonella', 'Escherichia']))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(34954, 13)"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Taxonomic Annotation(Genus Level)'].str.contains('|'.join(['Salmonella', 'Escherichia']))].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.095501</td>\n",
       "      <td>0.096262</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.098658</td>\n",
       "      <td>0.104673</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN;USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>754082.ECSTEC7V_0573</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.068666</td>\n",
       "      <td>0.073832</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285627</th>\n",
       "      <td>656444.ECNG_05213</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285628</th>\n",
       "      <td>656444.ECNG_04574</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.042620</td>\n",
       "      <td>0.049533</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285629</th>\n",
       "      <td>656444.ECNG_05118</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>COG1475</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285638</th>\n",
       "      <td>753642.ECNC101_09444</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285641</th>\n",
       "      <td>926026.ECO5101_18612</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>proNOG05789</td>\n",
       "      <td>0.001579</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34954 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "1               585054.EFER_0542        21669          Complete            SP   \n",
       "2             SZEY-48A_GL0052647        20778          Complete           CHN   \n",
       "3        1048689.ECO55CA74_02930        20778          Complete            SP   \n",
       "4               MH0427_GL0087973        20775          Complete           EUR   \n",
       "5           754082.ECSTEC7V_0573        20775          Complete            SP   \n",
       "...                          ...          ...               ...           ...   \n",
       "1285627        656444.ECNG_05213          102          Complete            SP   \n",
       "1285628        656444.ECNG_04574          102          Complete            SP   \n",
       "1285629        656444.ECNG_05118          102          Complete            SP   \n",
       "1285638     753642.ECNC101_09444          102          Complete            SP   \n",
       "1285641     926026.ECO5101_18612          102          Complete            SP   \n",
       "\n",
       "        Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "1                           Proteobacteria                       Escherichia   \n",
       "2                           Proteobacteria                       Escherichia   \n",
       "3                           Proteobacteria                       Escherichia   \n",
       "4                           Proteobacteria                       Escherichia   \n",
       "5                           Proteobacteria                       Escherichia   \n",
       "...                                    ...                               ...   \n",
       "1285627                     Proteobacteria                       Escherichia   \n",
       "1285628                     Proteobacteria                       Escherichia   \n",
       "1285629                     Proteobacteria                       Escherichia   \n",
       "1285638                     Proteobacteria                       Escherichia   \n",
       "1285641                     Proteobacteria                       Escherichia   \n",
       "\n",
       "        KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "1               unknown          NOG12793                    0.058406   \n",
       "2                K01317          NOG12793                    0.203631   \n",
       "3                K01317          NOG12793                    0.095501   \n",
       "4                K01317          NOG12793                    0.098658   \n",
       "5                K01317          NOG12793                    0.068666   \n",
       "...                 ...               ...                         ...   \n",
       "1285627         unknown           unknown                    0.000789   \n",
       "1285628         unknown           unknown                    0.042620   \n",
       "1285629         unknown           COG1475                    0.000000   \n",
       "1285638         unknown           unknown                    0.001579   \n",
       "1285641         unknown       proNOG05789                    0.001579   \n",
       "\n",
       "         Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "1                              0.062617                    unknown   \n",
       "2                              0.212150            Enzyme Families   \n",
       "3                              0.096262            Enzyme Families   \n",
       "4                              0.104673            Enzyme Families   \n",
       "5                              0.073832            Enzyme Families   \n",
       "...                                 ...                        ...   \n",
       "1285627                        0.000935                    unknown   \n",
       "1285628                        0.049533                    unknown   \n",
       "1285629                        0.000000                    unknown   \n",
       "1285638                        0.000935                    unknown   \n",
       "1285641                        0.001869                    unknown   \n",
       "\n",
       "        eggNOG Functional Categories Cohort Assembled  \n",
       "1                   Function unknown          EUR;CHN  \n",
       "2                   Function unknown          EUR;CHN  \n",
       "3                   Function unknown          EUR;CHN  \n",
       "4                   Function unknown      EUR;CHN;USA  \n",
       "5                   Function unknown          EUR;CHN  \n",
       "...                              ...              ...  \n",
       "1285627                      unknown              NaN  \n",
       "1285628                      unknown              NaN  \n",
       "1285629                Transcription              NaN  \n",
       "1285638                      unknown              NaN  \n",
       "1285641             Function unknown              NaN  \n",
       "\n",
       "[34954 rows x 13 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterias_patog = ['Salmonella', 'Escherichia']\n",
    "df[df['Taxonomic Annotation(Genus Level)'].isin(bacterias_patog)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**El encadenamiento de métodos es un gran poder que conlleva una gran responsabilidad.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>K12548</th>\n",
       "      <td>457398.HMPREF0326_01805</td>\n",
       "      <td>15099</td>\n",
       "      <td>Cellular Processes and Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01179</th>\n",
       "      <td>360104.CCC13826_0899</td>\n",
       "      <td>15243</td>\n",
       "      <td>Carbohydrate Metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01317</th>\n",
       "      <td>386585.ECs0542</td>\n",
       "      <td>15876</td>\n",
       "      <td>Enzyme Families</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01447</th>\n",
       "      <td>556261.HMPREF0240_03368</td>\n",
       "      <td>16590</td>\n",
       "      <td>Cellular Processes and Signaling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K00599</th>\n",
       "      <td>MH0244_GL0138579</td>\n",
       "      <td>16611</td>\n",
       "      <td>Amino Acid Metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K14194</th>\n",
       "      <td>411490.ANACAC_02896</td>\n",
       "      <td>16887</td>\n",
       "      <td>Infectious Diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01779</th>\n",
       "      <td>684738.LLKF_1213</td>\n",
       "      <td>17136</td>\n",
       "      <td>Amino Acid Metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01179</th>\n",
       "      <td>MH0185_GL0087605</td>\n",
       "      <td>17184</td>\n",
       "      <td>Carbohydrate Metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K00599</th>\n",
       "      <td>ED50A_GL0021676</td>\n",
       "      <td>17310</td>\n",
       "      <td>Amino Acid Metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K15125</th>\n",
       "      <td>591001.Acfer_0201</td>\n",
       "      <td>17715</td>\n",
       "      <td>Infectious Diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01361</th>\n",
       "      <td>T2D-108A_GL0085403</td>\n",
       "      <td>18006</td>\n",
       "      <td>Enzyme Families</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K08218</th>\n",
       "      <td>1045856.EcWSU1_02111</td>\n",
       "      <td>18006</td>\n",
       "      <td>Membrane Transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01317</th>\n",
       "      <td>1001989.PPECC33_4430</td>\n",
       "      <td>18462</td>\n",
       "      <td>Enzyme Families</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01317</th>\n",
       "      <td>MH0315_GL0002178</td>\n",
       "      <td>18975</td>\n",
       "      <td>Enzyme Families</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01317</th>\n",
       "      <td>754082.ECSTEC7V_0573</td>\n",
       "      <td>20775</td>\n",
       "      <td>Enzyme Families</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01317</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Enzyme Families</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01317</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Enzyme Families</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K01317</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Enzyme Families</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Gene Name  Gene Length  \\\n",
       "KEGG Annotation                                         \n",
       "K12548           457398.HMPREF0326_01805        15099   \n",
       "K01179              360104.CCC13826_0899        15243   \n",
       "K01317                    386585.ECs0542        15876   \n",
       "K01447           556261.HMPREF0240_03368        16590   \n",
       "K00599                  MH0244_GL0138579        16611   \n",
       "K14194               411490.ANACAC_02896        16887   \n",
       "K01779                  684738.LLKF_1213        17136   \n",
       "K01179                  MH0185_GL0087605        17184   \n",
       "K00599                   ED50A_GL0021676        17310   \n",
       "K15125                 591001.Acfer_0201        17715   \n",
       "K01361                T2D-108A_GL0085403        18006   \n",
       "K08218              1045856.EcWSU1_02111        18006   \n",
       "K01317              1001989.PPECC33_4430        18462   \n",
       "K01317                  MH0315_GL0002178        18975   \n",
       "K01317              754082.ECSTEC7V_0573        20775   \n",
       "K01317                  MH0427_GL0087973        20775   \n",
       "K01317           1048689.ECO55CA74_02930        20778   \n",
       "K01317                SZEY-48A_GL0052647        20778   \n",
       "\n",
       "                       KEGG Functional Categories  \n",
       "KEGG Annotation                                    \n",
       "K12548           Cellular Processes and Signaling  \n",
       "K01179                    Carbohydrate Metabolism  \n",
       "K01317                            Enzyme Families  \n",
       "K01447           Cellular Processes and Signaling  \n",
       "K00599                      Amino Acid Metabolism  \n",
       "K14194                        Infectious Diseases  \n",
       "K01779                      Amino Acid Metabolism  \n",
       "K01179                    Carbohydrate Metabolism  \n",
       "K00599                      Amino Acid Metabolism  \n",
       "K15125                        Infectious Diseases  \n",
       "K01361                            Enzyme Families  \n",
       "K08218                         Membrane Transport  \n",
       "K01317                            Enzyme Families  \n",
       "K01317                            Enzyme Families  \n",
       "K01317                            Enzyme Families  \n",
       "K01317                            Enzyme Families  \n",
       "K01317                            Enzyme Families  \n",
       "K01317                            Enzyme Families  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df.loc[((df['Gene Length'] >= 15000) & \n",
    "         (df['Gene Completeness'] >= 'Complete') &\n",
    "         (df['KEGG Annotation'] != 'unknown')), \n",
    "       ['Gene Name', 'KEGG Annotation', 'Gene Length', \n",
    "        'KEGG Functional Categories']]\n",
    "    .sort_values('Gene Length')\n",
    "    .set_index('KEGG Annotation'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selección de campos en *DataFrames*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La selección de campos en DataFrames, *slicing*, puede utilizar tres diferentes sintaxis:\n",
    "- `df[]`\n",
    "- `df.loc[]`\n",
    "- `df.iloc[]`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`loc` es para *location*, y `iloc` para *index location*\n",
    "\n",
    "|Operación| `df[]` | `df.loc[]`| `df.iloc[]`|\n",
    "| :--- | :--- | :--- |:---|\n",
    "|Seleccione una sola columna por etiqueta|`df['A']`|`df.loc[:, 'A']`| `-` |\n",
    "|Seleccionar lista de columnas por etiqueta|`df[['A', 'C']]`|`df.loc[:, ['A', 'C']]`|`-`|\n",
    "|Cortar columnas por etiqueta|`-`|`df.loc[:, 'A':'C']`|`-`|\n",
    "|Seleccione una sola columna por posición|`-`|`-`|`df.iloc[:, 1]`|\n",
    "|Seleccionar lista de columnas por posición|`-`|`-`|`df.iloc[:, [0, 2]]`|\n",
    "|Cortar columnas por posición|`-`|`-`|`df.iloc[:, 0:2]`|\n",
    "|Seleccione una sola fila por etiqueta|`-`|`df.loc['a']*`|`-`|\n",
    "|Seleccione una lista de filas por etiqueta|`-`|`df.loc[['a', 'b']]*`|`-`|\n",
    "|Cortar filas por etiqueta|`df['a':'d']*`|`df.loc['b':'d']*`|`-`|\n",
    "|Seleccione una sola fila por posición|`-`|`-`|`df.iloc[1]`|\n",
    "|Seleccionar una lista de filas por posición|`-`|`-`|`df.iloc[[1, 3]]`|\n",
    "|Cortar filas por posición|`df[1:4]`|`-`|`df.iloc[1:4]`|\n",
    "|Seleccionar lista de filas y columnas por etiqueta|`-`|`df.loc[['b', 'c'], ['A', 'C']]*`|`-`|\n",
    "|Seleccionar lista de filas y columnas por posición|`-`|`-`|`df.iloc[[1, 3], [2, 1]]`|\n",
    "|Cortar filas y columnas por etiqueta|`-`|`df.locp[['b': 'c'], ['A': 'C']]*`|`-`|\n",
    "|Cortar filas y columnas por posición|`-`|`-`|`df.iloc[1:3, 0:2]`|\n",
    "\n",
    "**\\*** El índice de la fila debe ser *string*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hemos seleccionado columnas usando el índice de columnas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    911104.WcibK1_010100007220\n",
       "1              585054.EFER_0542\n",
       "2            SZEY-48A_GL0052647\n",
       "3       1048689.ECO55CA74_02930\n",
       "4              MH0427_GL0087973\n",
       "Name: Gene Name, dtype: object"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Gene Name'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>Gene Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>unknown</td>\n",
       "      <td>24615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>unknown</td>\n",
       "      <td>21669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>K01317</td>\n",
       "      <td>20778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>K01317</td>\n",
       "      <td>20778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>K01317</td>\n",
       "      <td>20775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene Name KEGG Annotation  Gene Length\n",
       "0  911104.WcibK1_010100007220         unknown        24615\n",
       "1            585054.EFER_0542         unknown        21669\n",
       "2          SZEY-48A_GL0052647          K01317        20778\n",
       "3     1048689.ECO55CA74_02930          K01317        20778\n",
       "4            MH0427_GL0087973          K01317        20775"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Gene Name', 'KEGG Annotation', 'Gene Length']].head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "El *slicing* permite obtener la localización basado en un índice numérico o en el índice de la columna."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "      <th>Taxonomic Annotation(Genus Level)</th>\n",
       "      <th>KEGG Annotation</th>\n",
       "      <th>eggNOG Annotation</th>\n",
       "      <th>Sample Occurence Frequency</th>\n",
       "      <th>Individual Occurence Frequency</th>\n",
       "      <th>KEGG Functional Categories</th>\n",
       "      <th>eggNOG Functional Categories</th>\n",
       "      <th>Cohort Assembled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>24615</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Weissella</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.066298</td>\n",
       "      <td>0.072897</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.058406</td>\n",
       "      <td>0.062617</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>K01317</td>\n",
       "      <td>NOG12793</td>\n",
       "      <td>0.203631</td>\n",
       "      <td>0.212150</td>\n",
       "      <td>Enzyme Families</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>EUR;CHN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene Name  Gene Length Gene Completeness Cohort Origin  \\\n",
       "0  911104.WcibK1_010100007220        24615          Complete            SP   \n",
       "1            585054.EFER_0542        21669          Complete            SP   \n",
       "2          SZEY-48A_GL0052647        20778          Complete           CHN   \n",
       "\n",
       "  Taxonomic Annotation(Phylum Level) Taxonomic Annotation(Genus Level)  \\\n",
       "0                         Firmicutes                         Weissella   \n",
       "1                     Proteobacteria                       Escherichia   \n",
       "2                     Proteobacteria                       Escherichia   \n",
       "\n",
       "  KEGG Annotation eggNOG Annotation  Sample Occurence Frequency  \\\n",
       "0         unknown          NOG12793                    0.066298   \n",
       "1         unknown          NOG12793                    0.058406   \n",
       "2          K01317          NOG12793                    0.203631   \n",
       "\n",
       "   Individual Occurence Frequency KEGG Functional Categories  \\\n",
       "0                        0.072897                    unknown   \n",
       "1                        0.062617                    unknown   \n",
       "2                        0.212150            Enzyme Families   \n",
       "\n",
       "  eggNOG Functional Categories Cohort Assembled  \n",
       "0             Function unknown              EUR  \n",
       "1             Function unknown          EUR;CHN  \n",
       "2             Function unknown          EUR;CHN  "
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[0:3].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    911104.WcibK1_010100007220\n",
       "1              585054.EFER_0542\n",
       "2            SZEY-48A_GL0052647\n",
       "3       1048689.ECO55CA74_02930\n",
       "4              MH0427_GL0087973\n",
       "Name: Gene Name, dtype: object"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[0:4, 'Gene Name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          911104.WcibK1_010100007220\n",
       "1                    585054.EFER_0542\n",
       "2                  SZEY-48A_GL0052647\n",
       "3             1048689.ECO55CA74_02930\n",
       "4                    MH0427_GL0087973\n",
       "                      ...            \n",
       "1285637       665954.HMPREF1017_02554\n",
       "1285638          753642.ECNC101_09444\n",
       "1285639        768724.HMPREF9131_0965\n",
       "1285640        866771.HMPREF9296_1721\n",
       "1285641          926026.ECO5101_18612\n",
       "Name: Gene Name, Length: 1285642, dtype: object"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[:, 'Gene Name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    911104.WcibK1_010100007220\n",
       "1              585054.EFER_0542\n",
       "2            SZEY-48A_GL0052647\n",
       "3       1048689.ECO55CA74_02930\n",
       "4              MH0427_GL0087973\n",
       "Name: Gene Name, dtype: object"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[:, 'Gene Name'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>24615</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285637</th>\n",
       "      <td>665954.HMPREF1017_02554</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285638</th>\n",
       "      <td>753642.ECNC101_09444</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285639</th>\n",
       "      <td>768724.HMPREF9131_0965</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285640</th>\n",
       "      <td>866771.HMPREF9296_1721</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285641</th>\n",
       "      <td>926026.ECO5101_18612</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1285642 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Gene Name  Gene Length Gene Completeness  \\\n",
       "0        911104.WcibK1_010100007220        24615          Complete   \n",
       "1                  585054.EFER_0542        21669          Complete   \n",
       "2                SZEY-48A_GL0052647        20778          Complete   \n",
       "3           1048689.ECO55CA74_02930        20778          Complete   \n",
       "4                  MH0427_GL0087973        20775          Complete   \n",
       "...                             ...          ...               ...   \n",
       "1285637     665954.HMPREF1017_02554          102          Complete   \n",
       "1285638        753642.ECNC101_09444          102          Complete   \n",
       "1285639      768724.HMPREF9131_0965          102          Complete   \n",
       "1285640      866771.HMPREF9296_1721          102          Complete   \n",
       "1285641        926026.ECO5101_18612          102          Complete   \n",
       "\n",
       "        Cohort Origin  \n",
       "0                  SP  \n",
       "1                  SP  \n",
       "2                 CHN  \n",
       "3                  SP  \n",
       "4                 EUR  \n",
       "...               ...  \n",
       "1285637            SP  \n",
       "1285638            SP  \n",
       "1285639            SP  \n",
       "1285640            SP  \n",
       "1285641            SP  \n",
       "\n",
       "[1285642 rows x 4 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[:, 'Gene Name':'Cohort Origin']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>911104.WcibK1_010100007220</td>\n",
       "      <td>24615</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585054.EFER_0542</td>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SZEY-48A_GL0052647</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1048689.ECO55CA74_02930</td>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MH0427_GL0087973</td>\n",
       "      <td>20775</td>\n",
       "      <td>Complete</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285637</th>\n",
       "      <td>665954.HMPREF1017_02554</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285638</th>\n",
       "      <td>753642.ECNC101_09444</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285639</th>\n",
       "      <td>768724.HMPREF9131_0965</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285640</th>\n",
       "      <td>866771.HMPREF9296_1721</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285641</th>\n",
       "      <td>926026.ECO5101_18612</td>\n",
       "      <td>102</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1285642 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Gene Name  Gene Length Gene Completeness  \\\n",
       "0        911104.WcibK1_010100007220        24615          Complete   \n",
       "1                  585054.EFER_0542        21669          Complete   \n",
       "2                SZEY-48A_GL0052647        20778          Complete   \n",
       "3           1048689.ECO55CA74_02930        20778          Complete   \n",
       "4                  MH0427_GL0087973        20775          Complete   \n",
       "...                             ...          ...               ...   \n",
       "1285637     665954.HMPREF1017_02554          102          Complete   \n",
       "1285638        753642.ECNC101_09444          102          Complete   \n",
       "1285639      768724.HMPREF9131_0965          102          Complete   \n",
       "1285640      866771.HMPREF9296_1721          102          Complete   \n",
       "1285641        926026.ECO5101_18612          102          Complete   \n",
       "\n",
       "        Cohort Origin  \n",
       "0                  SP  \n",
       "1                  SP  \n",
       "2                 CHN  \n",
       "3                  SP  \n",
       "4                 EUR  \n",
       "...               ...  \n",
       "1285637            SP  \n",
       "1285638            SP  \n",
       "1285639            SP  \n",
       "1285640            SP  \n",
       "1285641            SP  \n",
       "\n",
       "[1285642 rows x 4 columns]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, 0:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Gene Completeness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>24615</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21669</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20778</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20778</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20775</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285637</th>\n",
       "      <td>102</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285638</th>\n",
       "      <td>102</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285639</th>\n",
       "      <td>102</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285640</th>\n",
       "      <td>102</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1285641</th>\n",
       "      <td>102</td>\n",
       "      <td>SP</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1285642 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene Length Cohort Origin Gene Completeness\n",
       "0              24615            SP          Complete\n",
       "1              21669            SP          Complete\n",
       "2              20778           CHN          Complete\n",
       "3              20778            SP          Complete\n",
       "4              20775           EUR          Complete\n",
       "...              ...           ...               ...\n",
       "1285637          102            SP          Complete\n",
       "1285638          102            SP          Complete\n",
       "1285639          102            SP          Complete\n",
       "1285640          102            SP          Complete\n",
       "1285641          102            SP          Complete\n",
       "\n",
       "[1285642 rows x 3 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [1, 3, 2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Length</th>\n",
       "      <th>Gene Completeness</th>\n",
       "      <th>Cohort Origin</th>\n",
       "      <th>Taxonomic Annotation(Phylum Level)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>24615</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Firmicutes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21669</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>CHN</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20778</td>\n",
       "      <td>Complete</td>\n",
       "      <td>SP</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene Length Gene Completeness Cohort Origin  \\\n",
       "0        24615          Complete            SP   \n",
       "1        21669          Complete            SP   \n",
       "2        20778          Complete           CHN   \n",
       "3        20778          Complete            SP   \n",
       "\n",
       "  Taxonomic Annotation(Phylum Level)  \n",
       "0                         Firmicutes  \n",
       "1                     Proteobacteria  \n",
       "2                     Proteobacteria  \n",
       "3                     Proteobacteria  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[0:4, 1:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "¿Y si quiero obtener un rango de columnas para hacer *slicing*?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pandas profiling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas_profiling import ProfileReport"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "profile = ProfileReport(df, title=\"Explorando con Pandas\", explorative=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "profile.to_file(\"reporte_IGC.html\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
